**AAN_new에 대하여 N,A1,A2로 가능한 모든 경우에 대해 에러데이터 생성.**
- AAN_new는 시계열로 자르되, 각 파일의 앞 20%를 테스트로 한다.
- 80%는 트레이닝으로 한다.
- 기존과 다른점은, 에러를 넣을 구간을 순차적으로 고르지않고, random하게 선택한다.
- 대부분의 데이터가 비슷한 값을 가지므로, 중간에 랜덤하게 삽압된 에러는 trend를 통해 걸러낼 수 있을 것으로 생각한다.


# **A1: P1.B3004** col=3
# **A2: P1.B3005** col=4
# **N: P1.B4002** col=5


In [2]:
!pip install ipython-autotime
%load_ext autotime
import pickle
import numpy as np
import pandas as pd
import os 
import random
import math

origin = pd.read_pickle(os.getcwd() + '/drive/My Drive/data/AAN_new/reset_NoError_whole_x_test.pickle')
origin


time  P1.B2004  ...  attack.P3  error
0       2019-11-01 20:00:00+09:00    0.0998  ...          0      0
1       2019-11-01 20:00:01+09:00    0.0998  ...          0      0
2       2019-11-01 20:00:02+09:00    0.0998  ...          0      0
3       2019-11-01 20:00:03+09:00    0.0998  ...          0      0
4       2019-11-01 20:00:04+09:00    0.0998  ...          0      0
...                           ...       ...  ...        ...    ...
181795  2019-10-30 03:11:55+09:00    0.0305  ...          0      0
181796  2019-10-30 03:11:56+09:00    0.0305  ...          0      0
181797  2019-10-30 03:11:57+09:00    0.0305  ...          0      0
181798  2019-10-30 03:11:58+09:00    0.0305  ...          0      0
181799  2019-10-30 03:11:59+09:00    0.0305  ...          0      0

[181800 rows x 65 columns]

# **normal의 갯수 파악**

In [9]:
num0=0 # 노말
num1=0 # 어택
for i in origin["attack"]:
  if(i == 0):
    num0 = num0+1
  elif(i == 1):
    num1 = num1+1

print('normal 수')
print(num0)
print('attack 수')
print(num1)

normal 수
152888
attack 수
28912
time: 42.6 ms


# **테스트: 181800개**
# **노말:152888**
# **어택:28912**

# **기존방법의 문제**
- 위에서부터 순차적으로 에러를 넣었기 때문에, trend를 이용할 때 의미가 없음.
- 에러가 모여있으므로 계속 ---의 트렌드를 가지게된다.

# **에러생성 방법**
1. 데이터를 200개씩 909섹터로 나눈다.
1. 각 섹터의 normal만을 고른다.
2. 각 섹터의 normal의 갯수의 5%갯수만큼을 에러로한다.
3. 각 섹터의 normal의 인덱스를 갯수만큼 random하게 선택한다.
4. 선택된 row에 대하여, (원래의 값 + 섹터avg * 0.1) 값이 normal경계를 넘어서도록한다. 
5. 에러 플래그를 넣는다.

# **테스트용**

In [34]:
x_test = pd.read_pickle(os.getcwd() + '/drive/My Drive/data/AAN_new/reset_NoError_whole_x_test.pickle')
x_test.drop('time', axis=1, inplace=True)

indexNames = x_test[x_test['attack'] == 1 ].index
print('attack의 index 확인')
indexNames
# # Delete these row indexes from dataFrame
# a.drop(indexNames , inplace=True)


attack의 index 확인


Int64Index([110160, 110161, 110162, 110163, 110164, 110165, 110166, 110167,
            110168, 110169,
            ...
            143543, 143544, 143545, 143546, 143547, 143548, 143549, 143550,
            143551, 143552],
           dtype='int64', length=28912)

time: 180 ms


In [0]:
picklename = '/drive/My Drive/data/AAN_new/A1/reset_Error_whole_x_test_5percent.pickle'
percent = 0.05
target = 2
max = 500 



x_test = pd.read_pickle(os.getcwd() + '/drive/My Drive/data/AAN_new/reset_NoError_whole_x_test.pickle')
x_test.drop('time', axis=1, inplace=True)
avg_ratio = 0.1     # 평균의 반영비율
baesu = 1           # 평균*배수. 배수 설정.
sector_size = 200
sector_num = 909      

for i in range(0, sector_num):
  sector_start = i * sector_size
  sector_end = sector_start + sector_size

  temp = x_test.iloc[sector_start:sector_end]   # 섹터추출
  normal_index = temp.loc[(temp.attack == 0)].index # 섹터에서 normal인 index 추출
  # print('normal_index')
  # print(normal_index)
  # print('normal_index 길이')
  # print(len(normal_index))
  rand_index = random.sample(list(normal_index), int(len(normal_index)*percent))  # normal갯수*percent의 올림갯수만큼 random선택.   
  # print('rand_index')
  # print(rand_index)
  # print('rand_index 길이')
  # print(len(rand_index))
  # print("")

  # 타겟에 더할 값 계산. avg = 평균*비율. avg는 평균이 아님.
  avg = temp.iloc[:, target].mean(axis=0) * avg_ratio
  # print(avg)
  # print(x_test.iloc[normal_index, target])


  # random index를 가지고 x_test에 에러 넣기
  for j in (rand_index):
    new_value = 0
    times = baesu
    while(new_value < max + avg):
      new_value = x_test.iloc[j, target] + (avg*times)    # 에러의 세기 충분하도록 함
      times += 0.3
    
    x_test.iloc[j, target] = new_value  # 에러가 들어간 값으로 업데이트
    x_test.iloc[j, 63] = 1  # 에러플래그 1로

# x_test.head(50)

# 에러갯수 확인
num0=0
num1=0
for i in x_test["error"]:
  if(i==0):
    num0 = num0+1
  elif(i==1):
    num1 = num1+1
print('에러없는 값(attack+normal)')
print(num0)
print('노말에러')
print(num1)

############################################### 저장 구간 ##################################################
# save as pickle here.
x_test.to_pickle(os.getcwd() + picklename)
print("pickle saved.")

# pickle test
a = pd.read_pickle(os.getcwd() + picklename)
print("")
print("로드된 데이터")
num0=0
num1=0
for i in a["error"]:
  if(i==0):
    num0 = num0+1
  elif(i==1):
    num1 = num1+1

print('에러 없음')
print(num0)
print('에러 있음')
print(num1)


# Get names of indexes for which column Age has value 30
indexNames = a[a['error'] == 0 ].index
# Delete these row indexes from dataFrame
a.drop(indexNames , inplace=True)
a


# **[1] generate A1**
# col=3


# **싱글 5%**

In [0]:
picklename = '/drive/My Drive/data/AAN_new/A1/reset_Error_whole_x_test_5percent.pickle'
percent = 0.05
target = 2
max = 500 

# A1: 2 / 500
# A2: 3 / 1100
# N:  4 / 32


avg_ratio = 0.1     # 평균의 반영비율
baesu = 1           # 평균*배수. 배수 설정.
sector_size = 200
sector_num = 909     

x_test = pd.read_pickle(os.getcwd() + '/drive/My Drive/data/AAN_new/reset_NoError_whole_x_test.pickle')
x_test.drop('time', axis=1, inplace=True)
 
for i in range(0, sector_num):
  sector_start = i * sector_size
  sector_end = sector_start + sector_size

  temp = x_test.iloc[sector_start:sector_end]   # 섹터추출
  normal_index = temp.loc[(temp.attack == 0)].index # 섹터에서 normal인 index 추출
  rand_index = random.sample(list(normal_index), int(len(normal_index)*percent))  # normal갯수*percent의 올림갯수만큼 random선택.   

  # 타겟에 더할 값 계산. avg = 평균*비율. avg는 평균이 아님.
  avg = temp.iloc[:, target].mean(axis=0) * avg_ratio

  # random index를 가지고 x_test에 에러 넣기
  for j in (rand_index):
    new_value = 0
    times = baesu
    while(new_value < max + avg):
      new_value = x_test.iloc[j, target] + (avg*times)    # 에러의 세기 충분하도록 함
      times += 0.3
    x_test.iloc[j, target] = new_value  # 에러가 들어간 값으로 업데이트
    x_test.iloc[j, 63] = 1  # 에러플래그 1로

############################################### 저장 구간 ##################################################
# save as pickle here.
x_test.to_pickle(os.getcwd() + picklename)
print("pickle saved.")

# pickle test
a = pd.read_pickle(os.getcwd() + picklename)
print("")
print("로드된 데이터")
num0=0
num1=0
for i in a["error"]:
  if(i==0):
    num0 = num0+1
  elif(i==1):
    num1 = num1+1

print('에러 X')
print(num0)
print('에러 O')
print(num1)


# Get names of indexes for which column Age has value 30
indexNames = a[a['error'] == 0 ].index
# Delete these row indexes from dataFrame
a.drop(indexNames , inplace=True)
a


10

In [51]:
picklename = '/drive/My Drive/data/AAN_new/A1/reset_Error_whole_x_test_10percent.pickle'
percent = 0.1
target = 2
max = 500 

# A1: 2 / 500
# A2: 3 / 1100
# N:  4 / 32


avg_ratio = 0.1     # 평균의 반영비율
baesu = 1           # 평균*배수. 배수 설정.
sector_size = 200
sector_num = 909     

x_test = pd.read_pickle(os.getcwd() + '/drive/My Drive/data/AAN_new/reset_NoError_whole_x_test.pickle')
x_test.drop('time', axis=1, inplace=True)
 
for i in range(0, sector_num):
  sector_start = i * sector_size
  sector_end = sector_start + sector_size

  temp = x_test.iloc[sector_start:sector_end]   # 섹터추출
  normal_index = temp.loc[(temp.attack == 0)].index # 섹터에서 normal인 index 추출
  rand_index = random.sample(list(normal_index), int(len(normal_index)*percent))  # normal갯수*percent의 올림갯수만큼 random선택.   

  # 타겟에 더할 값 계산. avg = 평균*비율. avg는 평균이 아님.
  avg = temp.iloc[:, target].mean(axis=0) * avg_ratio

  # random index를 가지고 x_test에 에러 넣기
  for j in (rand_index):
    new_value = 0
    times = baesu
    while(new_value < max + avg):
      new_value = x_test.iloc[j, target] + (avg*times)    # 에러의 세기 충분하도록 함
      times += 0.3
    x_test.iloc[j, target] = new_value  # 에러가 들어간 값으로 업데이트
    x_test.iloc[j, 63] = 1  # 에러플래그 1로

############################################### 저장 구간 ##################################################
# save as pickle here.
x_test.to_pickle(os.getcwd() + picklename)
print("pickle saved.")

# pickle test
a = pd.read_pickle(os.getcwd() + picklename)
print("")
print("로드된 데이터")
num0=0
num1=0
for i in a["error"]:
  if(i==0):
    num0 = num0+1
  elif(i==1):
    num1 = num1+1

print('에러 X')
print(num0)
print('에러 O')
print(num1)


# Get names of indexes for which column Age has value 30
indexNames = a[a['error'] == 0 ].index
# Delete these row indexes from dataFrame
a.drop(indexNames , inplace=True)
a


pickle saved.

로드된 데이터
에러 X
166512
에러 O
15288


P1.B2004  P1.B2016    P1.B3004  ...  attack.P2  attack.P3  error
5         0.0998    1.3804  541.599731  ...          0          0      1
8         0.0998    1.3723  541.599731  ...          0          0      1
16        0.0998    1.3589  541.599731  ...          0          0      1
27        0.0998    1.3386  541.599731  ...          0          0      1
48        0.0998    1.3196  541.599731  ...          0          0      1
...          ...       ...         ...  ...        ...        ...    ...
181749    0.0305    1.3150  549.262512  ...          0          0      1
181752    0.0305    1.3296  549.262512  ...          0          0      1
181773    0.0305    1.3518  549.262512  ...          0          0      1
181790    0.0305    1.3263  549.262512  ...          0          0      1
181799    0.0305    1.3251  549.262512  ...          0          0      1

[15288 rows x 64 columns]

time: 32.9 s


15

In [52]:
picklename = '/drive/My Drive/data/AAN_new/A1/reset_Error_whole_x_test_15percent.pickle'
percent = 0.15
target = 2
max = 500 

# A1: 2 / 500
# A2: 3 / 1100
# N:  4 / 32


avg_ratio = 0.1     # 평균의 반영비율
baesu = 1           # 평균*배수. 배수 설정.
sector_size = 200
sector_num = 909     

x_test = pd.read_pickle(os.getcwd() + '/drive/My Drive/data/AAN_new/reset_NoError_whole_x_test.pickle')
x_test.drop('time', axis=1, inplace=True)
 
for i in range(0, sector_num):
  sector_start = i * sector_size
  sector_end = sector_start + sector_size

  temp = x_test.iloc[sector_start:sector_end]   # 섹터추출
  normal_index = temp.loc[(temp.attack == 0)].index # 섹터에서 normal인 index 추출
  rand_index = random.sample(list(normal_index), int(len(normal_index)*percent))  # normal갯수*percent의 올림갯수만큼 random선택.   

  # 타겟에 더할 값 계산. avg = 평균*비율. avg는 평균이 아님.
  avg = temp.iloc[:, target].mean(axis=0) * avg_ratio

  # random index를 가지고 x_test에 에러 넣기
  for j in (rand_index):
    new_value = 0
    times = baesu
    while(new_value < max + avg):
      new_value = x_test.iloc[j, target] + (avg*times)    # 에러의 세기 충분하도록 함
      times += 0.3
    x_test.iloc[j, target] = new_value  # 에러가 들어간 값으로 업데이트
    x_test.iloc[j, 63] = 1  # 에러플래그 1로

############################################### 저장 구간 ##################################################
# save as pickle here.
x_test.to_pickle(os.getcwd() + picklename)
print("pickle saved.")

# pickle test
a = pd.read_pickle(os.getcwd() + picklename)
print("")
print("로드된 데이터")
num0=0
num1=0
for i in a["error"]:
  if(i==0):
    num0 = num0+1
  elif(i==1):
    num1 = num1+1

print('에러 X')
print(num0)
print('에러 O')
print(num1)


# Get names of indexes for which column Age has value 30
indexNames = a[a['error'] == 0 ].index
# Delete these row indexes from dataFrame
a.drop(indexNames , inplace=True)
a


pickle saved.

로드된 데이터
에러 X
158867
에러 O
22933


P1.B2004  P1.B2016    P1.B3004  ...  attack.P2  attack.P3  error
0         0.0998    1.3901  541.599731  ...          0          0      1
1         0.0998    1.3840  541.599731  ...          0          0      1
23        0.0998    1.3456  541.599731  ...          0          0      1
25        0.0998    1.3413  541.599731  ...          0          0      1
38        0.0998    1.3233  541.599731  ...          0          0      1
...          ...       ...         ...  ...        ...        ...    ...
181764    0.0305    1.3896  549.262512  ...          0          0      1
181765    0.0305    1.3953  549.262512  ...          0          0      1
181782    0.0305    1.3042  549.262512  ...          0          0      1
181785    0.0305    1.3176  549.262512  ...          0          0      1
181797    0.0305    1.3254  549.262512  ...          0          0      1

[22933 rows x 64 columns]

time: 48.6 s


30

In [53]:
picklename = '/drive/My Drive/data/AAN_new/A1/reset_Error_whole_x_test_30percent.pickle'
percent = 0.3
target = 2
max = 500 

# A1: 2 / 500
# A2: 3 / 1100
# N:  4 / 32


avg_ratio = 0.1     # 평균의 반영비율
baesu = 1           # 평균*배수. 배수 설정.
sector_size = 200
sector_num = 909     

x_test = pd.read_pickle(os.getcwd() + '/drive/My Drive/data/AAN_new/reset_NoError_whole_x_test.pickle')
x_test.drop('time', axis=1, inplace=True)
 
for i in range(0, sector_num):
  sector_start = i * sector_size
  sector_end = sector_start + sector_size

  temp = x_test.iloc[sector_start:sector_end]   # 섹터추출
  normal_index = temp.loc[(temp.attack == 0)].index # 섹터에서 normal인 index 추출
  rand_index = random.sample(list(normal_index), int(len(normal_index)*percent))  # normal갯수*percent의 올림갯수만큼 random선택.   

  # 타겟에 더할 값 계산. avg = 평균*비율. avg는 평균이 아님.
  avg = temp.iloc[:, target].mean(axis=0) * avg_ratio

  # random index를 가지고 x_test에 에러 넣기
  for j in (rand_index):
    new_value = 0
    times = baesu
    while(new_value < max + avg):
      new_value = x_test.iloc[j, target] + (avg*times)    # 에러의 세기 충분하도록 함
      times += 0.3
    x_test.iloc[j, target] = new_value  # 에러가 들어간 값으로 업데이트
    x_test.iloc[j, 63] = 1  # 에러플래그 1로

############################################### 저장 구간 ##################################################
# save as pickle here.
x_test.to_pickle(os.getcwd() + picklename)
print("pickle saved.")

# pickle test
a = pd.read_pickle(os.getcwd() + picklename)
print("")
print("로드된 데이터")
num0=0
num1=0
for i in a["error"]:
  if(i==0):
    num0 = num0+1
  elif(i==1):
    num1 = num1+1

print('에러 X')
print(num0)
print('에러 O')
print(num1)


# Get names of indexes for which column Age has value 30
indexNames = a[a['error'] == 0 ].index
# Delete these row indexes from dataFrame
a.drop(indexNames , inplace=True)
a


pickle saved.

로드된 데이터
에러 X
135934
에러 O
45866


P1.B2004  P1.B2016    P1.B3004  ...  attack.P2  attack.P3  error
2         0.0998    1.3837  541.599731  ...          0          0      1
5         0.0998    1.3804  541.599731  ...          0          0      1
7         0.0998    1.3746  541.599731  ...          0          0      1
9         0.0998    1.3711  541.599731  ...          0          0      1
12        0.0998    1.3647  541.599731  ...          0          0      1
...          ...       ...         ...  ...        ...        ...    ...
181784    0.0305    1.3129  549.262512  ...          0          0      1
181788    0.0305    1.3252  549.262512  ...          0          0      1
181789    0.0305    1.3185  549.262512  ...          0          0      1
181791    0.0305    1.3232  549.262512  ...          0          0      1
181796    0.0305    1.3258  549.262512  ...          0          0      1

[45866 rows x 64 columns]

time: 1min 33s


# **A2**

5

In [3]:
picklename = '/drive/My Drive/data/AAN_new/A2/reset_Error_whole_x_test_5percent.pickle'
percent = 0.05
target = 3
max = 1100

# A1: 2 / 500
# A2: 3 / 1100
# N:  4 / 32


avg_ratio = 0.1     # 평균의 반영비율
baesu = 1           # 평균*배수. 배수 설정.
sector_size = 200
sector_num = 909     

x_test = pd.read_pickle(os.getcwd() + '/drive/My Drive/data/AAN_new/reset_NoError_whole_x_test.pickle')
x_test.drop('time', axis=1, inplace=True)
 
for i in range(0, sector_num):
  sector_start = i * sector_size
  sector_end = sector_start + sector_size

  temp = x_test.iloc[sector_start:sector_end]   # 섹터추출
  normal_index = temp.loc[(temp.attack == 0)].index # 섹터에서 normal인 index 추출
  rand_index = random.sample(list(normal_index), int(len(normal_index)*percent))  # normal갯수*percent의 올림갯수만큼 random선택.   

  # 타겟에 더할 값 계산. avg = 평균*비율. avg는 평균이 아님.
  avg = temp.iloc[:, target].mean(axis=0) * avg_ratio

  # random index를 가지고 x_test에 에러 넣기
  for j in (rand_index):
    new_value = 0
    times = baesu
    while(new_value < max + avg):
      new_value = x_test.iloc[j, target] + (avg*times)    # 에러의 세기 충분하도록 함
      times += 0.3
    x_test.iloc[j, target] = new_value  # 에러가 들어간 값으로 업데이트
    x_test.iloc[j, 63] = 1  # 에러플래그 1로

############################################### 저장 구간 ##################################################
# save as pickle here.
x_test.to_pickle(os.getcwd() + picklename)
print("pickle saved.")

# pickle test
a = pd.read_pickle(os.getcwd() + picklename)
print("")
print("로드된 데이터")
num0=0
num1=0
for i in a["error"]:
  if(i==0):
    num0 = num0+1
  elif(i==1):
    num1 = num1+1

print('에러 X')
print(num0)
print('에러 O')
print(num1)


# Get names of indexes for which column Age has value 30
indexNames = a[a['error'] == 0 ].index
# Delete these row indexes from dataFrame
a.drop(indexNames , inplace=True)
a


pickle saved.

로드된 데이터
에러 X
174156
에러 O
7644


P1.B2004  P1.B2016    P1.B3004  ...  attack.P2  attack.P3  error
38        0.0998    1.3233  404.178711  ...          0          0      1
40        0.0998    1.3216  404.178711  ...          0          0      1
65        0.0998    1.4480  404.178711  ...          0          0      1
69        0.0998    1.4809  404.178711  ...          0          0      1
78        0.0998    1.5459  404.178711  ...          0          0      1
...          ...       ...         ...  ...        ...        ...    ...
181744    0.0305    1.2688  392.330292  ...          0          0      1
181763    0.0305    1.3859  392.330292  ...          0          0      1
181768    0.0305    1.3896  392.330292  ...          0          0      1
181769    0.0305    1.3814  392.330292  ...          0          0      1
181776    0.0305    1.3292  392.330292  ...          0          0      1

[7644 rows x 64 columns]

time: 16.5 s


10

In [4]:
picklename = '/drive/My Drive/data/AAN_new/A2/reset_Error_whole_x_test_10percent.pickle'
percent = 0.1
target = 3
max = 1100

# A1: 2 / 500
# A2: 3 / 1100
# N:  4 / 32


avg_ratio = 0.1     # 평균의 반영비율
baesu = 1           # 평균*배수. 배수 설정.
sector_size = 200
sector_num = 909     

x_test = pd.read_pickle(os.getcwd() + '/drive/My Drive/data/AAN_new/reset_NoError_whole_x_test.pickle')
x_test.drop('time', axis=1, inplace=True)
 
for i in range(0, sector_num):
  sector_start = i * sector_size
  sector_end = sector_start + sector_size

  temp = x_test.iloc[sector_start:sector_end]   # 섹터추출
  normal_index = temp.loc[(temp.attack == 0)].index # 섹터에서 normal인 index 추출
  rand_index = random.sample(list(normal_index), int(len(normal_index)*percent))  # normal갯수*percent의 올림갯수만큼 random선택.   

  # 타겟에 더할 값 계산. avg = 평균*비율. avg는 평균이 아님.
  avg = temp.iloc[:, target].mean(axis=0) * avg_ratio

  # random index를 가지고 x_test에 에러 넣기
  for j in (rand_index):
    new_value = 0
    times = baesu
    while(new_value < max + avg):
      new_value = x_test.iloc[j, target] + (avg*times)    # 에러의 세기 충분하도록 함
      times += 0.3
    x_test.iloc[j, target] = new_value  # 에러가 들어간 값으로 업데이트
    x_test.iloc[j, 63] = 1  # 에러플래그 1로

############################################### 저장 구간 ##################################################
# save as pickle here.
x_test.to_pickle(os.getcwd() + picklename)
print("pickle saved.")

# pickle test
a = pd.read_pickle(os.getcwd() + picklename)
print("")
print("로드된 데이터")
num0=0
num1=0
for i in a["error"]:
  if(i==0):
    num0 = num0+1
  elif(i==1):
    num1 = num1+1

print('에러 X')
print(num0)
print('에러 O')
print(num1)


# Get names of indexes for which column Age has value 30
indexNames = a[a['error'] == 0 ].index
# Delete these row indexes from dataFrame
a.drop(indexNames , inplace=True)
a


pickle saved.

로드된 데이터
에러 X
166512
에러 O
15288


P1.B2004  P1.B2016    P1.B3004  ...  attack.P2  attack.P3  error
0         0.0998    1.3901  404.178711  ...          0          0      1
4         0.0998    1.3839  404.178711  ...          0          0      1
22        0.0998    1.3453  404.178711  ...          0          0      1
25        0.0998    1.3413  404.178711  ...          0          0      1
27        0.0998    1.3386  404.178711  ...          0          0      1
...          ...       ...         ...  ...        ...        ...    ...
181736    0.0305    1.1891  392.330292  ...          0          0      1
181739    0.0305    1.2189  392.330292  ...          0          0      1
181746    0.0305    1.2853  392.330292  ...          0          0      1
181768    0.0305    1.3896  392.330292  ...          0          0      1
181783    0.0305    1.3083  392.330292  ...          0          0      1

[15288 rows x 64 columns]

time: 31.8 s


15

In [5]:
picklename = '/drive/My Drive/data/AAN_new/A2/reset_Error_whole_x_test_15percent.pickle'
percent = 0.15
target = 3
max = 1100

# A1: 2 / 500
# A2: 3 / 1100
# N:  4 / 32


avg_ratio = 0.1     # 평균의 반영비율
baesu = 1           # 평균*배수. 배수 설정.
sector_size = 200
sector_num = 909     

x_test = pd.read_pickle(os.getcwd() + '/drive/My Drive/data/AAN_new/reset_NoError_whole_x_test.pickle')
x_test.drop('time', axis=1, inplace=True)
 
for i in range(0, sector_num):
  sector_start = i * sector_size
  sector_end = sector_start + sector_size

  temp = x_test.iloc[sector_start:sector_end]   # 섹터추출
  normal_index = temp.loc[(temp.attack == 0)].index # 섹터에서 normal인 index 추출
  rand_index = random.sample(list(normal_index), int(len(normal_index)*percent))  # normal갯수*percent의 올림갯수만큼 random선택.   

  # 타겟에 더할 값 계산. avg = 평균*비율. avg는 평균이 아님.
  avg = temp.iloc[:, target].mean(axis=0) * avg_ratio

  # random index를 가지고 x_test에 에러 넣기
  for j in (rand_index):
    new_value = 0
    times = baesu
    while(new_value < max + avg):
      new_value = x_test.iloc[j, target] + (avg*times)    # 에러의 세기 충분하도록 함
      times += 0.3
    x_test.iloc[j, target] = new_value  # 에러가 들어간 값으로 업데이트
    x_test.iloc[j, 63] = 1  # 에러플래그 1로

############################################### 저장 구간 ##################################################
# save as pickle here.
x_test.to_pickle(os.getcwd() + picklename)
print("pickle saved.")

# pickle test
a = pd.read_pickle(os.getcwd() + picklename)
print("")
print("로드된 데이터")
num0=0
num1=0
for i in a["error"]:
  if(i==0):
    num0 = num0+1
  elif(i==1):
    num1 = num1+1

print('에러 X')
print(num0)
print('에러 O')
print(num1)


# Get names of indexes for which column Age has value 30
indexNames = a[a['error'] == 0 ].index
# Delete these row indexes from dataFrame
a.drop(indexNames , inplace=True)
a


pickle saved.

로드된 데이터
에러 X
158867
에러 O
22933


P1.B2004  P1.B2016    P1.B3004  ...  attack.P2  attack.P3  error
11        0.0998    1.3647  404.178711  ...          0          0      1
17        0.0998    1.3564  404.178711  ...          0          0      1
18        0.0998    1.3528  404.178711  ...          0          0      1
25        0.0998    1.3413  404.178711  ...          0          0      1
44        0.0998    1.3181  404.178711  ...          0          0      1
...          ...       ...         ...  ...        ...        ...    ...
181785    0.0305    1.3176  392.330292  ...          0          0      1
181786    0.0305    1.3254  392.330292  ...          0          0      1
181788    0.0305    1.3252  392.330292  ...          0          0      1
181791    0.0305    1.3232  392.330292  ...          0          0      1
181796    0.0305    1.3258  392.330292  ...          0          0      1

[22933 rows x 64 columns]

time: 45 s


30

In [6]:
picklename = '/drive/My Drive/data/AAN_new/A2/reset_Error_whole_x_test_30percent.pickle'
percent = 0.3
target = 3
max = 1100

# A1: 2 / 500
# A2: 3 / 1100
# N:  4 / 32


avg_ratio = 0.1     # 평균의 반영비율
baesu = 1           # 평균*배수. 배수 설정.
sector_size = 200
sector_num = 909     

x_test = pd.read_pickle(os.getcwd() + '/drive/My Drive/data/AAN_new/reset_NoError_whole_x_test.pickle')
x_test.drop('time', axis=1, inplace=True)
 
for i in range(0, sector_num):
  sector_start = i * sector_size
  sector_end = sector_start + sector_size

  temp = x_test.iloc[sector_start:sector_end]   # 섹터추출
  normal_index = temp.loc[(temp.attack == 0)].index # 섹터에서 normal인 index 추출
  rand_index = random.sample(list(normal_index), int(len(normal_index)*percent))  # normal갯수*percent의 올림갯수만큼 random선택.   

  # 타겟에 더할 값 계산. avg = 평균*비율. avg는 평균이 아님.
  avg = temp.iloc[:, target].mean(axis=0) * avg_ratio

  # random index를 가지고 x_test에 에러 넣기
  for j in (rand_index):
    new_value = 0
    times = baesu
    while(new_value < max + avg):
      new_value = x_test.iloc[j, target] + (avg*times)    # 에러의 세기 충분하도록 함
      times += 0.3
    x_test.iloc[j, target] = new_value  # 에러가 들어간 값으로 업데이트
    x_test.iloc[j, 63] = 1  # 에러플래그 1로

############################################### 저장 구간 ##################################################
# save as pickle here.
x_test.to_pickle(os.getcwd() + picklename)
print("pickle saved.")

# pickle test
a = pd.read_pickle(os.getcwd() + picklename)
print("")
print("로드된 데이터")
num0=0
num1=0
for i in a["error"]:
  if(i==0):
    num0 = num0+1
  elif(i==1):
    num1 = num1+1

print('에러 X')
print(num0)
print('에러 O')
print(num1)


# Get names of indexes for which column Age has value 30
indexNames = a[a['error'] == 0 ].index
# Delete these row indexes from dataFrame
a.drop(indexNames , inplace=True)
a


pickle saved.

로드된 데이터
에러 X
135934
에러 O
45866


P1.B2004  P1.B2016    P1.B3004  ...  attack.P2  attack.P3  error
3         0.0998    1.3854  404.178711  ...          0          0      1
5         0.0998    1.3804  404.178711  ...          0          0      1
6         0.0998    1.3785  404.178711  ...          0          0      1
12        0.0998    1.3647  404.178711  ...          0          0      1
13        0.0998    1.3642  404.178711  ...          0          0      1
...          ...       ...         ...  ...        ...        ...    ...
181775    0.0305    1.3383  392.330292  ...          0          0      1
181779    0.0305    1.2974  392.330292  ...          0          0      1
181780    0.0305    1.2997  392.330292  ...          0          0      1
181781    0.0305    1.3051  392.330292  ...          0          0      1
181794    0.0305    1.3209  392.330292  ...          0          0      1

[45866 rows x 64 columns]

time: 1min 29s


# **N**

5

In [7]:
picklename = '/drive/My Drive/data/AAN_new/N/reset_Error_whole_x_test_5percent.pickle'
percent = 0.05
target = 4
max = 32 

# A1: 2 / 500
# A2: 3 / 1100
# N:  4 / 32


avg_ratio = 0.1     # 평균의 반영비율
baesu = 1           # 평균*배수. 배수 설정.
sector_size = 200
sector_num = 909     

x_test = pd.read_pickle(os.getcwd() + '/drive/My Drive/data/AAN_new/reset_NoError_whole_x_test.pickle')
x_test.drop('time', axis=1, inplace=True)
 
for i in range(0, sector_num):
  sector_start = i * sector_size
  sector_end = sector_start + sector_size

  temp = x_test.iloc[sector_start:sector_end]   # 섹터추출
  normal_index = temp.loc[(temp.attack == 0)].index # 섹터에서 normal인 index 추출
  rand_index = random.sample(list(normal_index), int(len(normal_index)*percent))  # normal갯수*percent의 올림갯수만큼 random선택.   

  # 타겟에 더할 값 계산. avg = 평균*비율. avg는 평균이 아님.
  avg = temp.iloc[:, target].mean(axis=0) * avg_ratio

  # random index를 가지고 x_test에 에러 넣기
  for j in (rand_index):
    new_value = 0
    times = baesu
    while(new_value < max + avg):
      new_value = x_test.iloc[j, target] + (avg*times)    # 에러의 세기 충분하도록 함
      times += 0.3
    x_test.iloc[j, target] = new_value  # 에러가 들어간 값으로 업데이트
    x_test.iloc[j, 63] = 1  # 에러플래그 1로

############################################### 저장 구간 ##################################################
# save as pickle here.
x_test.to_pickle(os.getcwd() + picklename)
print("pickle saved.")

# pickle test
a = pd.read_pickle(os.getcwd() + picklename)
print("")
print("로드된 데이터")
num0=0
num1=0
for i in a["error"]:
  if(i==0):
    num0 = num0+1
  elif(i==1):
    num1 = num1+1

print('에러 X')
print(num0)
print('에러 O')
print(num1)


# Get names of indexes for which column Age has value 30
indexNames = a[a['error'] == 0 ].index
# Delete these row indexes from dataFrame
a.drop(indexNames , inplace=True)
a


pickle saved.

로드된 데이터
에러 X
174156
에러 O
7644


P1.B2004  P1.B2016    P1.B3004  ...  attack.P2  attack.P3  error
32        0.0998    1.3312  404.178711  ...          0          0      1
38        0.0998    1.3233  404.178711  ...          0          0      1
56        0.0998    1.3682  404.178711  ...          0          0      1
84        0.0998    1.5758  404.178711  ...          0          0      1
100       0.0998    1.6399  404.178711  ...          0          0      1
...          ...       ...         ...  ...        ...        ...    ...
181669    0.0305    1.3400  392.330292  ...          0          0      1
181702    0.0305    1.1685  392.330292  ...          0          0      1
181766    0.0305    1.3992  392.330292  ...          0          0      1
181770    0.0305    1.3777  392.330292  ...          0          0      1
181794    0.0305    1.3209  392.330292  ...          0          0      1

[7644 rows x 64 columns]

time: 16.9 s


10

In [8]:
picklename = '/drive/My Drive/data/AAN_new/N/reset_Error_whole_x_test_10percent.pickle'
percent = 0.1
target = 4
max = 32 

# A1: 2 / 500
# A2: 3 / 1100
# N:  4 / 32


avg_ratio = 0.1     # 평균의 반영비율
baesu = 1           # 평균*배수. 배수 설정.
sector_size = 200
sector_num = 909     

x_test = pd.read_pickle(os.getcwd() + '/drive/My Drive/data/AAN_new/reset_NoError_whole_x_test.pickle')
x_test.drop('time', axis=1, inplace=True)
 
for i in range(0, sector_num):
  sector_start = i * sector_size
  sector_end = sector_start + sector_size

  temp = x_test.iloc[sector_start:sector_end]   # 섹터추출
  normal_index = temp.loc[(temp.attack == 0)].index # 섹터에서 normal인 index 추출
  rand_index = random.sample(list(normal_index), int(len(normal_index)*percent))  # normal갯수*percent의 올림갯수만큼 random선택.   

  # 타겟에 더할 값 계산. avg = 평균*비율. avg는 평균이 아님.
  avg = temp.iloc[:, target].mean(axis=0) * avg_ratio

  # random index를 가지고 x_test에 에러 넣기
  for j in (rand_index):
    new_value = 0
    times = baesu
    while(new_value < max + avg):
      new_value = x_test.iloc[j, target] + (avg*times)    # 에러의 세기 충분하도록 함
      times += 0.3
    x_test.iloc[j, target] = new_value  # 에러가 들어간 값으로 업데이트
    x_test.iloc[j, 63] = 1  # 에러플래그 1로

############################################### 저장 구간 ##################################################
# save as pickle here.
x_test.to_pickle(os.getcwd() + picklename)
print("pickle saved.")

# pickle test
a = pd.read_pickle(os.getcwd() + picklename)
print("")
print("로드된 데이터")
num0=0
num1=0
for i in a["error"]:
  if(i==0):
    num0 = num0+1
  elif(i==1):
    num1 = num1+1

print('에러 X')
print(num0)
print('에러 O')
print(num1)


# Get names of indexes for which column Age has value 30
indexNames = a[a['error'] == 0 ].index
# Delete these row indexes from dataFrame
a.drop(indexNames , inplace=True)
a


pickle saved.

로드된 데이터
에러 X
166512
에러 O
15288


P1.B2004  P1.B2016    P1.B3004  ...  attack.P2  attack.P3  error
1         0.0998    1.3840  404.178711  ...          0          0      1
15        0.0998    1.3621  404.178711  ...          0          0      1
26        0.0998    1.3380  404.178711  ...          0          0      1
41        0.0998    1.3207  404.178711  ...          0          0      1
53        0.0998    1.3393  404.178711  ...          0          0      1
...          ...       ...         ...  ...        ...        ...    ...
181744    0.0305    1.2688  392.330292  ...          0          0      1
181762    0.0305    1.3788  392.330292  ...          0          0      1
181772    0.0305    1.3560  392.330292  ...          0          0      1
181781    0.0305    1.3051  392.330292  ...          0          0      1
181799    0.0305    1.3251  392.330292  ...          0          0      1

[15288 rows x 64 columns]

time: 32.9 s


15

In [9]:
picklename = '/drive/My Drive/data/AAN_new/N/reset_Error_whole_x_test_15percent.pickle'
percent = 0.15
target = 4
max = 32 

# A1: 2 / 500
# A2: 3 / 1100
# N:  4 / 32


avg_ratio = 0.1     # 평균의 반영비율
baesu = 1           # 평균*배수. 배수 설정.
sector_size = 200
sector_num = 909     

x_test = pd.read_pickle(os.getcwd() + '/drive/My Drive/data/AAN_new/reset_NoError_whole_x_test.pickle')
x_test.drop('time', axis=1, inplace=True)
 
for i in range(0, sector_num):
  sector_start = i * sector_size
  sector_end = sector_start + sector_size

  temp = x_test.iloc[sector_start:sector_end]   # 섹터추출
  normal_index = temp.loc[(temp.attack == 0)].index # 섹터에서 normal인 index 추출
  rand_index = random.sample(list(normal_index), int(len(normal_index)*percent))  # normal갯수*percent의 올림갯수만큼 random선택.   

  # 타겟에 더할 값 계산. avg = 평균*비율. avg는 평균이 아님.
  avg = temp.iloc[:, target].mean(axis=0) * avg_ratio

  # random index를 가지고 x_test에 에러 넣기
  for j in (rand_index):
    new_value = 0
    times = baesu
    while(new_value < max + avg):
      new_value = x_test.iloc[j, target] + (avg*times)    # 에러의 세기 충분하도록 함
      times += 0.3
    x_test.iloc[j, target] = new_value  # 에러가 들어간 값으로 업데이트
    x_test.iloc[j, 63] = 1  # 에러플래그 1로

############################################### 저장 구간 ##################################################
# save as pickle here.
x_test.to_pickle(os.getcwd() + picklename)
print("pickle saved.")

# pickle test
a = pd.read_pickle(os.getcwd() + picklename)
print("")
print("로드된 데이터")
num0=0
num1=0
for i in a["error"]:
  if(i==0):
    num0 = num0+1
  elif(i==1):
    num1 = num1+1

print('에러 X')
print(num0)
print('에러 O')
print(num1)


# Get names of indexes for which column Age has value 30
indexNames = a[a['error'] == 0 ].index
# Delete these row indexes from dataFrame
a.drop(indexNames , inplace=True)
a


pickle saved.

로드된 데이터
에러 X
158867
에러 O
22933


P1.B2004  P1.B2016    P1.B3004  ...  attack.P2  attack.P3  error
2         0.0998    1.3837  404.178711  ...          0          0      1
3         0.0998    1.3854  404.178711  ...          0          0      1
6         0.0998    1.3785  404.178711  ...          0          0      1
8         0.0998    1.3723  404.178711  ...          0          0      1
10        0.0998    1.3688  404.178711  ...          0          0      1
...          ...       ...         ...  ...        ...        ...    ...
181779    0.0305    1.2974  392.330292  ...          0          0      1
181783    0.0305    1.3083  392.330292  ...          0          0      1
181788    0.0305    1.3252  392.330292  ...          0          0      1
181791    0.0305    1.3232  392.330292  ...          0          0      1
181799    0.0305    1.3251  392.330292  ...          0          0      1

[22933 rows x 64 columns]

time: 45.9 s


30

In [10]:
picklename = '/drive/My Drive/data/AAN_new/N/reset_Error_whole_x_test_30percent.pickle'
percent = 0.3
target = 4
max = 32 

# A1: 2 / 500
# A2: 3 / 1100
# N:  4 / 32


avg_ratio = 0.1     # 평균의 반영비율
baesu = 1           # 평균*배수. 배수 설정.
sector_size = 200
sector_num = 909     

x_test = pd.read_pickle(os.getcwd() + '/drive/My Drive/data/AAN_new/reset_NoError_whole_x_test.pickle')
x_test.drop('time', axis=1, inplace=True)
 
for i in range(0, sector_num):
  sector_start = i * sector_size
  sector_end = sector_start + sector_size

  temp = x_test.iloc[sector_start:sector_end]   # 섹터추출
  normal_index = temp.loc[(temp.attack == 0)].index # 섹터에서 normal인 index 추출
  rand_index = random.sample(list(normal_index), int(len(normal_index)*percent))  # normal갯수*percent의 올림갯수만큼 random선택.   

  # 타겟에 더할 값 계산. avg = 평균*비율. avg는 평균이 아님.
  avg = temp.iloc[:, target].mean(axis=0) * avg_ratio

  # random index를 가지고 x_test에 에러 넣기
  for j in (rand_index):
    new_value = 0
    times = baesu
    while(new_value < max + avg):
      new_value = x_test.iloc[j, target] + (avg*times)    # 에러의 세기 충분하도록 함
      times += 0.3
    x_test.iloc[j, target] = new_value  # 에러가 들어간 값으로 업데이트
    x_test.iloc[j, 63] = 1  # 에러플래그 1로

############################################### 저장 구간 ##################################################
# save as pickle here.
x_test.to_pickle(os.getcwd() + picklename)
print("pickle saved.")

# pickle test
a = pd.read_pickle(os.getcwd() + picklename)
print("")
print("로드된 데이터")
num0=0
num1=0
for i in a["error"]:
  if(i==0):
    num0 = num0+1
  elif(i==1):
    num1 = num1+1

print('에러 X')
print(num0)
print('에러 O')
print(num1)


# Get names of indexes for which column Age has value 30
indexNames = a[a['error'] == 0 ].index
# Delete these row indexes from dataFrame
a.drop(indexNames , inplace=True)
a


pickle saved.

로드된 데이터
에러 X
135934
에러 O
45866


P1.B2004  P1.B2016    P1.B3004  ...  attack.P2  attack.P3  error
4         0.0998    1.3839  404.178711  ...          0          0      1
11        0.0998    1.3647  404.178711  ...          0          0      1
13        0.0998    1.3642  404.178711  ...          0          0      1
21        0.0998    1.3495  404.178711  ...          0          0      1
25        0.0998    1.3413  404.178711  ...          0          0      1
...          ...       ...         ...  ...        ...        ...    ...
181785    0.0305    1.3176  392.330292  ...          0          0      1
181786    0.0305    1.3254  392.330292  ...          0          0      1
181791    0.0305    1.3232  392.330292  ...          0          0      1
181797    0.0305    1.3254  392.330292  ...          0          0      1
181798    0.0305    1.3295  392.330292  ...          0          0      1

[45866 rows x 64 columns]

time: 1min 29s


# **A1 N**

# **5%**

In [11]:
picklename = '/drive/My Drive/data/AAN_new/A1_N/reset_Error_whole_x_test_5percent.pickle'
percent = 0.05
target = 2
max = 500 
target2 = 4
max2 = 32

# A1: 2 / 500
# A2: 3 / 1100
# N:  4 / 32

avg_ratio = 0.1     # 평균의 반영비율
baesu = 1           # 평균*배수. 배수 설정.
sector_size = 200
sector_num = 909     

x_test = pd.read_pickle(os.getcwd() + '/drive/My Drive/data/AAN_new/reset_NoError_whole_x_test.pickle')
x_test.drop('time', axis=1, inplace=True)
 
for i in range(0, sector_num):
  sector_start = i * sector_size
  sector_end = sector_start + sector_size

  temp = x_test.iloc[sector_start:sector_end]   # 섹터추출
  normal_index = temp.loc[(temp.attack == 0)].index # 섹터에서 normal인 index 추출
  rand_index = random.sample(list(normal_index), int(len(normal_index)*percent))  # normal갯수*percent의 올림갯수만큼 random선택.   

  # 타겟에 더할 값 계산. avg = 평균*비율. avg는 평균이 아님.
  avg = temp.iloc[:, target].mean(axis=0) * avg_ratio
  avg2 = temp.iloc[:, target2].mean(axis=0) * avg_ratio
  

  # random index를 가지고 x_test에 에러 넣기
  for j in (rand_index):
    new_value = 0
    times = baesu
    while(new_value < max + avg):
      new_value = x_test.iloc[j, target] + (avg*times)    # 에러의 세기 충분하도록 함
      times += 0.3

    new_value2 = 0
    times = baesu
    while(new_value2 < max2 + avg2):
      new_value2 = x_test.iloc[j, target2] + (avg2*times)    # 에러의 세기 충분하도록 함
      times += 0.3
    
    x_test.iloc[j, target] = new_value  # 에러가 들어간 값으로 업데이트
    x_test.iloc[j, target2] = new_value2  # 에러가 들어간 값으로 업데이트
    x_test.iloc[j, 63] = 1  # 에러플래그 1로

############################################### 저장 구간 ##################################################
# save as pickle here.
x_test.to_pickle(os.getcwd() + picklename)
print("pickle saved.")

# pickle test
a = pd.read_pickle(os.getcwd() + picklename)
print("")
print("로드된 데이터")
num0=0
num1=0
for i in a["error"]:
  if(i==0):
    num0 = num0+1
  elif(i==1):
    num1 = num1+1

print('에러 X')
print(num0)
print('에러 O')
print(num1)


# Get names of indexes for which column Age has value 30
indexNames = a[a['error'] == 0 ].index
# Delete these row indexes from dataFrame
a.drop(indexNames , inplace=True)
a


pickle saved.

로드된 데이터
에러 X
174156
에러 O
7644


P1.B2004  P1.B2016    P1.B3004  ...  attack.P2  attack.P3  error
17        0.0998    1.3564  541.599731  ...          0          0      1
21        0.0998    1.3495  541.599731  ...          0          0      1
22        0.0998    1.3453  541.599731  ...          0          0      1
29        0.0998    1.3393  541.599731  ...          0          0      1
97        0.0998    1.6259  541.599731  ...          0          0      1
...          ...       ...         ...  ...        ...        ...    ...
181719    0.0305    1.0849  549.262512  ...          0          0      1
181724    0.0305    1.0702  549.262512  ...          0          0      1
181726    0.0305    1.0873  549.262512  ...          0          0      1
181745    0.0305    1.2791  549.262512  ...          0          0      1
181796    0.0305    1.3258  549.262512  ...          0          0      1

[7644 rows x 64 columns]

time: 25 s


10

In [12]:
picklename = '/drive/My Drive/data/AAN_new/A1_N/reset_Error_whole_x_test_10percent.pickle'
percent = 0.1
target = 2
max = 500 
target2 = 4
max2 = 32

# A1: 2 / 500
# A2: 3 / 1100
# N:  4 / 32

avg_ratio = 0.1     # 평균의 반영비율
baesu = 1           # 평균*배수. 배수 설정.
sector_size = 200
sector_num = 909     

x_test = pd.read_pickle(os.getcwd() + '/drive/My Drive/data/AAN_new/reset_NoError_whole_x_test.pickle')
x_test.drop('time', axis=1, inplace=True)
 
for i in range(0, sector_num):
  sector_start = i * sector_size
  sector_end = sector_start + sector_size

  temp = x_test.iloc[sector_start:sector_end]   # 섹터추출
  normal_index = temp.loc[(temp.attack == 0)].index # 섹터에서 normal인 index 추출
  rand_index = random.sample(list(normal_index), int(len(normal_index)*percent))  # normal갯수*percent의 올림갯수만큼 random선택.   

  # 타겟에 더할 값 계산. avg = 평균*비율. avg는 평균이 아님.
  avg = temp.iloc[:, target].mean(axis=0) * avg_ratio
  avg2 = temp.iloc[:, target2].mean(axis=0) * avg_ratio
  

  # random index를 가지고 x_test에 에러 넣기
  for j in (rand_index):
    new_value = 0
    times = baesu
    while(new_value < max + avg):
      new_value = x_test.iloc[j, target] + (avg*times)    # 에러의 세기 충분하도록 함
      times += 0.3

    new_value2 = 0
    times = baesu
    while(new_value2 < max2 + avg2):
      new_value2 = x_test.iloc[j, target2] + (avg2*times)    # 에러의 세기 충분하도록 함
      times += 0.3
    
    x_test.iloc[j, target] = new_value  # 에러가 들어간 값으로 업데이트
    x_test.iloc[j, target2] = new_value2  # 에러가 들어간 값으로 업데이트
    x_test.iloc[j, 63] = 1  # 에러플래그 1로

############################################### 저장 구간 ##################################################
# save as pickle here.
x_test.to_pickle(os.getcwd() + picklename)
print("pickle saved.")

# pickle test
a = pd.read_pickle(os.getcwd() + picklename)
print("")
print("로드된 데이터")
num0=0
num1=0
for i in a["error"]:
  if(i==0):
    num0 = num0+1
  elif(i==1):
    num1 = num1+1

print('에러 X')
print(num0)
print('에러 O')
print(num1)


# Get names of indexes for which column Age has value 30
indexNames = a[a['error'] == 0 ].index
# Delete these row indexes from dataFrame
a.drop(indexNames , inplace=True)
a


pickle saved.

로드된 데이터
에러 X
166512
에러 O
15288


P1.B2004  P1.B2016    P1.B3004  ...  attack.P2  attack.P3  error
0         0.0998    1.3901  541.599731  ...          0          0      1
10        0.0998    1.3688  541.599731  ...          0          0      1
12        0.0998    1.3647  541.599731  ...          0          0      1
33        0.0998    1.3337  541.599731  ...          0          0      1
56        0.0998    1.3682  541.599731  ...          0          0      1
...          ...       ...         ...  ...        ...        ...    ...
181755    0.0305    1.3438  549.262512  ...          0          0      1
181759    0.0305    1.3600  549.262512  ...          0          0      1
181760    0.0305    1.3692  549.262512  ...          0          0      1
181768    0.0305    1.3896  549.262512  ...          0          0      1
181782    0.0305    1.3042  549.262512  ...          0          0      1

[15288 rows x 64 columns]

time: 49.4 s


15

In [13]:
picklename = '/drive/My Drive/data/AAN_new/A1_N/reset_Error_whole_x_test_15percent.pickle'
percent = 0.15
target = 2
max = 500 
target2 = 4
max2 = 32

# A1: 2 / 500
# A2: 3 / 1100
# N:  4 / 32

avg_ratio = 0.1     # 평균의 반영비율
baesu = 1           # 평균*배수. 배수 설정.
sector_size = 200
sector_num = 909     

x_test = pd.read_pickle(os.getcwd() + '/drive/My Drive/data/AAN_new/reset_NoError_whole_x_test.pickle')
x_test.drop('time', axis=1, inplace=True)
 
for i in range(0, sector_num):
  sector_start = i * sector_size
  sector_end = sector_start + sector_size

  temp = x_test.iloc[sector_start:sector_end]   # 섹터추출
  normal_index = temp.loc[(temp.attack == 0)].index # 섹터에서 normal인 index 추출
  rand_index = random.sample(list(normal_index), int(len(normal_index)*percent))  # normal갯수*percent의 올림갯수만큼 random선택.   

  # 타겟에 더할 값 계산. avg = 평균*비율. avg는 평균이 아님.
  avg = temp.iloc[:, target].mean(axis=0) * avg_ratio
  avg2 = temp.iloc[:, target2].mean(axis=0) * avg_ratio
  

  # random index를 가지고 x_test에 에러 넣기
  for j in (rand_index):
    new_value = 0
    times = baesu
    while(new_value < max + avg):
      new_value = x_test.iloc[j, target] + (avg*times)    # 에러의 세기 충분하도록 함
      times += 0.3

    new_value2 = 0
    times = baesu
    while(new_value2 < max2 + avg2):
      new_value2 = x_test.iloc[j, target2] + (avg2*times)    # 에러의 세기 충분하도록 함
      times += 0.3
    
    x_test.iloc[j, target] = new_value  # 에러가 들어간 값으로 업데이트
    x_test.iloc[j, target2] = new_value2  # 에러가 들어간 값으로 업데이트
    x_test.iloc[j, 63] = 1  # 에러플래그 1로

############################################### 저장 구간 ##################################################
# save as pickle here.
x_test.to_pickle(os.getcwd() + picklename)
print("pickle saved.")

# pickle test
a = pd.read_pickle(os.getcwd() + picklename)
print("")
print("로드된 데이터")
num0=0
num1=0
for i in a["error"]:
  if(i==0):
    num0 = num0+1
  elif(i==1):
    num1 = num1+1

print('에러 X')
print(num0)
print('에러 O')
print(num1)


# Get names of indexes for which column Age has value 30
indexNames = a[a['error'] == 0 ].index
# Delete these row indexes from dataFrame
a.drop(indexNames , inplace=True)
a


pickle saved.

로드된 데이터
에러 X
158867
에러 O
22933


P1.B2004  P1.B2016    P1.B3004  ...  attack.P2  attack.P3  error
2         0.0998    1.3837  541.599731  ...          0          0      1
8         0.0998    1.3723  541.599731  ...          0          0      1
14        0.0998    1.3636  541.599731  ...          0          0      1
16        0.0998    1.3589  541.599731  ...          0          0      1
27        0.0998    1.3386  541.599731  ...          0          0      1
...          ...       ...         ...  ...        ...        ...    ...
181756    0.0305    1.3536  549.262512  ...          0          0      1
181776    0.0305    1.3292  549.262512  ...          0          0      1
181784    0.0305    1.3129  549.262512  ...          0          0      1
181786    0.0305    1.3254  549.262512  ...          0          0      1
181787    0.0305    1.3270  549.262512  ...          0          0      1

[22933 rows x 64 columns]

time: 1min 10s


30

In [14]:
picklename = '/drive/My Drive/data/AAN_new/A1_N/reset_Error_whole_x_test_30percent.pickle'
percent = 0.3
target = 2
max = 500 
target2 = 4
max2 = 32

# A1: 2 / 500
# A2: 3 / 1100
# N:  4 / 32

avg_ratio = 0.1     # 평균의 반영비율
baesu = 1           # 평균*배수. 배수 설정.
sector_size = 200
sector_num = 909     

x_test = pd.read_pickle(os.getcwd() + '/drive/My Drive/data/AAN_new/reset_NoError_whole_x_test.pickle')
x_test.drop('time', axis=1, inplace=True)
 
for i in range(0, sector_num):
  sector_start = i * sector_size
  sector_end = sector_start + sector_size

  temp = x_test.iloc[sector_start:sector_end]   # 섹터추출
  normal_index = temp.loc[(temp.attack == 0)].index # 섹터에서 normal인 index 추출
  rand_index = random.sample(list(normal_index), int(len(normal_index)*percent))  # normal갯수*percent의 올림갯수만큼 random선택.   

  # 타겟에 더할 값 계산. avg = 평균*비율. avg는 평균이 아님.
  avg = temp.iloc[:, target].mean(axis=0) * avg_ratio
  avg2 = temp.iloc[:, target2].mean(axis=0) * avg_ratio
  

  # random index를 가지고 x_test에 에러 넣기
  for j in (rand_index):
    new_value = 0
    times = baesu
    while(new_value < max + avg):
      new_value = x_test.iloc[j, target] + (avg*times)    # 에러의 세기 충분하도록 함
      times += 0.3

    new_value2 = 0
    times = baesu
    while(new_value2 < max2 + avg2):
      new_value2 = x_test.iloc[j, target2] + (avg2*times)    # 에러의 세기 충분하도록 함
      times += 0.3
    
    x_test.iloc[j, target] = new_value  # 에러가 들어간 값으로 업데이트
    x_test.iloc[j, target2] = new_value2  # 에러가 들어간 값으로 업데이트
    x_test.iloc[j, 63] = 1  # 에러플래그 1로

############################################### 저장 구간 ##################################################
# save as pickle here.
x_test.to_pickle(os.getcwd() + picklename)
print("pickle saved.")

# pickle test
a = pd.read_pickle(os.getcwd() + picklename)
print("")
print("로드된 데이터")
num0=0
num1=0
for i in a["error"]:
  if(i==0):
    num0 = num0+1
  elif(i==1):
    num1 = num1+1

print('에러 X')
print(num0)
print('에러 O')
print(num1)


# Get names of indexes for which column Age has value 30
indexNames = a[a['error'] == 0 ].index
# Delete these row indexes from dataFrame
a.drop(indexNames , inplace=True)
a


pickle saved.

로드된 데이터
에러 X
135934
에러 O
45866


P1.B2004  P1.B2016    P1.B3004  ...  attack.P2  attack.P3  error
0         0.0998    1.3901  541.599731  ...          0          0      1
1         0.0998    1.3840  541.599731  ...          0          0      1
3         0.0998    1.3854  541.599731  ...          0          0      1
5         0.0998    1.3804  541.599731  ...          0          0      1
6         0.0998    1.3785  541.599731  ...          0          0      1
...          ...       ...         ...  ...        ...        ...    ...
181788    0.0305    1.3252  549.262512  ...          0          0      1
181790    0.0305    1.3263  549.262512  ...          0          0      1
181792    0.0305    1.3269  549.262512  ...          0          0      1
181797    0.0305    1.3254  549.262512  ...          0          0      1
181798    0.0305    1.3295  549.262512  ...          0          0      1

[45866 rows x 64 columns]

time: 2min 15s


# **A2_N**

5

In [15]:
picklename = '/drive/My Drive/data/AAN_new/A2_N/reset_Error_whole_x_test_5percent.pickle'
percent = 0.05
target = 3
max = 1100 
target2 = 4
max2 = 32

# A1: 2 / 500
# A2: 3 / 1100
# N:  4 / 32

avg_ratio = 0.1     # 평균의 반영비율
baesu = 1           # 평균*배수. 배수 설정.
sector_size = 200
sector_num = 909     

x_test = pd.read_pickle(os.getcwd() + '/drive/My Drive/data/AAN_new/reset_NoError_whole_x_test.pickle')
x_test.drop('time', axis=1, inplace=True)
 
for i in range(0, sector_num):
  sector_start = i * sector_size
  sector_end = sector_start + sector_size

  temp = x_test.iloc[sector_start:sector_end]   # 섹터추출
  normal_index = temp.loc[(temp.attack == 0)].index # 섹터에서 normal인 index 추출
  rand_index = random.sample(list(normal_index), int(len(normal_index)*percent))  # normal갯수*percent의 올림갯수만큼 random선택.   

  # 타겟에 더할 값 계산. avg = 평균*비율. avg는 평균이 아님.
  avg = temp.iloc[:, target].mean(axis=0) * avg_ratio
  avg2 = temp.iloc[:, target2].mean(axis=0) * avg_ratio
  

  # random index를 가지고 x_test에 에러 넣기
  for j in (rand_index):
    new_value = 0
    times = baesu
    while(new_value < max + avg):
      new_value = x_test.iloc[j, target] + (avg*times)    # 에러의 세기 충분하도록 함
      times += 0.3

    new_value2 = 0
    times = baesu
    while(new_value2 < max2 + avg2):
      new_value2 = x_test.iloc[j, target2] + (avg2*times)    # 에러의 세기 충분하도록 함
      times += 0.3
    
    x_test.iloc[j, target] = new_value  # 에러가 들어간 값으로 업데이트
    x_test.iloc[j, target2] = new_value2  # 에러가 들어간 값으로 업데이트
    x_test.iloc[j, 63] = 1  # 에러플래그 1로

############################################### 저장 구간 ##################################################
# save as pickle here.
x_test.to_pickle(os.getcwd() + picklename)
print("pickle saved.")

# pickle test
a = pd.read_pickle(os.getcwd() + picklename)
print("")
print("로드된 데이터")
num0=0
num1=0
for i in a["error"]:
  if(i==0):
    num0 = num0+1
  elif(i==1):
    num1 = num1+1

print('에러 X')
print(num0)
print('에러 O')
print(num1)


# Get names of indexes for which column Age has value 30
indexNames = a[a['error'] == 0 ].index
# Delete these row indexes from dataFrame
a.drop(indexNames , inplace=True)
a


pickle saved.

로드된 데이터
에러 X
174156
에러 O
7644


P1.B2004  P1.B2016    P1.B3004  ...  attack.P2  attack.P3  error
31        0.0998    1.3341  404.178711  ...          0          0      1
38        0.0998    1.3233  404.178711  ...          0          0      1
82        0.0998    1.5659  404.178711  ...          0          0      1
89        0.0998    1.6013  404.178711  ...          0          0      1
100       0.0998    1.6399  404.178711  ...          0          0      1
...          ...       ...         ...  ...        ...        ...    ...
181758    0.0305    1.3574  392.330292  ...          0          0      1
181770    0.0305    1.3777  392.330292  ...          0          0      1
181778    0.0305    1.3074  392.330292  ...          0          0      1
181783    0.0305    1.3083  392.330292  ...          0          0      1
181787    0.0305    1.3270  392.330292  ...          0          0      1

[7644 rows x 64 columns]

time: 23.9 s


10

In [16]:
picklename = '/drive/My Drive/data/AAN_new/A2_N/reset_Error_whole_x_test_10percent.pickle'
percent = 0.1
target = 3
max = 1100 
target2 = 4
max2 = 32

# A1: 2 / 500
# A2: 3 / 1100
# N:  4 / 32

avg_ratio = 0.1     # 평균의 반영비율
baesu = 1           # 평균*배수. 배수 설정.
sector_size = 200
sector_num = 909     

x_test = pd.read_pickle(os.getcwd() + '/drive/My Drive/data/AAN_new/reset_NoError_whole_x_test.pickle')
x_test.drop('time', axis=1, inplace=True)
 
for i in range(0, sector_num):
  sector_start = i * sector_size
  sector_end = sector_start + sector_size

  temp = x_test.iloc[sector_start:sector_end]   # 섹터추출
  normal_index = temp.loc[(temp.attack == 0)].index # 섹터에서 normal인 index 추출
  rand_index = random.sample(list(normal_index), int(len(normal_index)*percent))  # normal갯수*percent의 올림갯수만큼 random선택.   

  # 타겟에 더할 값 계산. avg = 평균*비율. avg는 평균이 아님.
  avg = temp.iloc[:, target].mean(axis=0) * avg_ratio
  avg2 = temp.iloc[:, target2].mean(axis=0) * avg_ratio
  

  # random index를 가지고 x_test에 에러 넣기
  for j in (rand_index):
    new_value = 0
    times = baesu
    while(new_value < max + avg):
      new_value = x_test.iloc[j, target] + (avg*times)    # 에러의 세기 충분하도록 함
      times += 0.3

    new_value2 = 0
    times = baesu
    while(new_value2 < max2 + avg2):
      new_value2 = x_test.iloc[j, target2] + (avg2*times)    # 에러의 세기 충분하도록 함
      times += 0.3
    
    x_test.iloc[j, target] = new_value  # 에러가 들어간 값으로 업데이트
    x_test.iloc[j, target2] = new_value2  # 에러가 들어간 값으로 업데이트
    x_test.iloc[j, 63] = 1  # 에러플래그 1로

############################################### 저장 구간 ##################################################
# save as pickle here.
x_test.to_pickle(os.getcwd() + picklename)
print("pickle saved.")

# pickle test
a = pd.read_pickle(os.getcwd() + picklename)
print("")
print("로드된 데이터")
num0=0
num1=0
for i in a["error"]:
  if(i==0):
    num0 = num0+1
  elif(i==1):
    num1 = num1+1

print('에러 X')
print(num0)
print('에러 O')
print(num1)


# Get names of indexes for which column Age has value 30
indexNames = a[a['error'] == 0 ].index
# Delete these row indexes from dataFrame
a.drop(indexNames , inplace=True)
a


pickle saved.

로드된 데이터
에러 X
166512
에러 O
15288


P1.B2004  P1.B2016    P1.B3004  ...  attack.P2  attack.P3  error
1         0.0998    1.3840  404.178711  ...          0          0      1
18        0.0998    1.3528  404.178711  ...          0          0      1
19        0.0998    1.3480  404.178711  ...          0          0      1
33        0.0998    1.3337  404.178711  ...          0          0      1
37        0.0998    1.3247  404.178711  ...          0          0      1
...          ...       ...         ...  ...        ...        ...    ...
181750    0.0305    1.3269  392.330292  ...          0          0      1
181754    0.0305    1.3342  392.330292  ...          0          0      1
181771    0.0305    1.3661  392.330292  ...          0          0      1
181782    0.0305    1.3042  392.330292  ...          0          0      1
181797    0.0305    1.3254  392.330292  ...          0          0      1

[15288 rows x 64 columns]

time: 47.4 s


15

In [17]:
picklename = '/drive/My Drive/data/AAN_new/A2_N/reset_Error_whole_x_test_15percent.pickle'
percent = 0.15
target = 3
max = 1100 
target2 = 4
max2 = 32

# A1: 2 / 500
# A2: 3 / 1100
# N:  4 / 32

avg_ratio = 0.1     # 평균의 반영비율
baesu = 1           # 평균*배수. 배수 설정.
sector_size = 200
sector_num = 909     

x_test = pd.read_pickle(os.getcwd() + '/drive/My Drive/data/AAN_new/reset_NoError_whole_x_test.pickle')
x_test.drop('time', axis=1, inplace=True)
 
for i in range(0, sector_num):
  sector_start = i * sector_size
  sector_end = sector_start + sector_size

  temp = x_test.iloc[sector_start:sector_end]   # 섹터추출
  normal_index = temp.loc[(temp.attack == 0)].index # 섹터에서 normal인 index 추출
  rand_index = random.sample(list(normal_index), int(len(normal_index)*percent))  # normal갯수*percent의 올림갯수만큼 random선택.   

  # 타겟에 더할 값 계산. avg = 평균*비율. avg는 평균이 아님.
  avg = temp.iloc[:, target].mean(axis=0) * avg_ratio
  avg2 = temp.iloc[:, target2].mean(axis=0) * avg_ratio
  

  # random index를 가지고 x_test에 에러 넣기
  for j in (rand_index):
    new_value = 0
    times = baesu
    while(new_value < max + avg):
      new_value = x_test.iloc[j, target] + (avg*times)    # 에러의 세기 충분하도록 함
      times += 0.3

    new_value2 = 0
    times = baesu
    while(new_value2 < max2 + avg2):
      new_value2 = x_test.iloc[j, target2] + (avg2*times)    # 에러의 세기 충분하도록 함
      times += 0.3
    
    x_test.iloc[j, target] = new_value  # 에러가 들어간 값으로 업데이트
    x_test.iloc[j, target2] = new_value2  # 에러가 들어간 값으로 업데이트
    x_test.iloc[j, 63] = 1  # 에러플래그 1로

############################################### 저장 구간 ##################################################
# save as pickle here.
x_test.to_pickle(os.getcwd() + picklename)
print("pickle saved.")

# pickle test
a = pd.read_pickle(os.getcwd() + picklename)
print("")
print("로드된 데이터")
num0=0
num1=0
for i in a["error"]:
  if(i==0):
    num0 = num0+1
  elif(i==1):
    num1 = num1+1

print('에러 X')
print(num0)
print('에러 O')
print(num1)


# Get names of indexes for which column Age has value 30
indexNames = a[a['error'] == 0 ].index
# Delete these row indexes from dataFrame
a.drop(indexNames , inplace=True)
a


pickle saved.

로드된 데이터
에러 X
158867
에러 O
22933


P1.B2004  P1.B2016    P1.B3004  ...  attack.P2  attack.P3  error
4         0.0998    1.3839  404.178711  ...          0          0      1
16        0.0998    1.3589  404.178711  ...          0          0      1
18        0.0998    1.3528  404.178711  ...          0          0      1
36        0.0998    1.3274  404.178711  ...          0          0      1
43        0.0998    1.3166  404.178711  ...          0          0      1
...          ...       ...         ...  ...        ...        ...    ...
181754    0.0305    1.3342  392.330292  ...          0          0      1
181757    0.0305    1.3577  392.330292  ...          0          0      1
181770    0.0305    1.3777  392.330292  ...          0          0      1
181792    0.0305    1.3269  392.330292  ...          0          0      1
181797    0.0305    1.3254  392.330292  ...          0          0      1

[22933 rows x 64 columns]

time: 1min 8s


30

In [18]:
picklename = '/drive/My Drive/data/AAN_new/A2_N/reset_Error_whole_x_test_30percent.pickle'
percent = 0.3
target = 3
max = 1100 
target2 = 4
max2 = 32

# A1: 2 / 500
# A2: 3 / 1100
# N:  4 / 32

avg_ratio = 0.1     # 평균의 반영비율
baesu = 1           # 평균*배수. 배수 설정.
sector_size = 200
sector_num = 909     

x_test = pd.read_pickle(os.getcwd() + '/drive/My Drive/data/AAN_new/reset_NoError_whole_x_test.pickle')
x_test.drop('time', axis=1, inplace=True)
 
for i in range(0, sector_num):
  sector_start = i * sector_size
  sector_end = sector_start + sector_size

  temp = x_test.iloc[sector_start:sector_end]   # 섹터추출
  normal_index = temp.loc[(temp.attack == 0)].index # 섹터에서 normal인 index 추출
  rand_index = random.sample(list(normal_index), int(len(normal_index)*percent))  # normal갯수*percent의 올림갯수만큼 random선택.   

  # 타겟에 더할 값 계산. avg = 평균*비율. avg는 평균이 아님.
  avg = temp.iloc[:, target].mean(axis=0) * avg_ratio
  avg2 = temp.iloc[:, target2].mean(axis=0) * avg_ratio
  

  # random index를 가지고 x_test에 에러 넣기
  for j in (rand_index):
    new_value = 0
    times = baesu
    while(new_value < max + avg):
      new_value = x_test.iloc[j, target] + (avg*times)    # 에러의 세기 충분하도록 함
      times += 0.3

    new_value2 = 0
    times = baesu
    while(new_value2 < max2 + avg2):
      new_value2 = x_test.iloc[j, target2] + (avg2*times)    # 에러의 세기 충분하도록 함
      times += 0.3
    
    x_test.iloc[j, target] = new_value  # 에러가 들어간 값으로 업데이트
    x_test.iloc[j, target2] = new_value2  # 에러가 들어간 값으로 업데이트
    x_test.iloc[j, 63] = 1  # 에러플래그 1로

############################################### 저장 구간 ##################################################
# save as pickle here.
x_test.to_pickle(os.getcwd() + picklename)
print("pickle saved.")

# pickle test
a = pd.read_pickle(os.getcwd() + picklename)
print("")
print("로드된 데이터")
num0=0
num1=0
for i in a["error"]:
  if(i==0):
    num0 = num0+1
  elif(i==1):
    num1 = num1+1

print('에러 X')
print(num0)
print('에러 O')
print(num1)


# Get names of indexes for which column Age has value 30
indexNames = a[a['error'] == 0 ].index
# Delete these row indexes from dataFrame
a.drop(indexNames , inplace=True)
a


pickle saved.

로드된 데이터
에러 X
135934
에러 O
45866


P1.B2004  P1.B2016    P1.B3004  ...  attack.P2  attack.P3  error
0         0.0998    1.3901  404.178711  ...          0          0      1
1         0.0998    1.3840  404.178711  ...          0          0      1
10        0.0998    1.3688  404.178711  ...          0          0      1
13        0.0998    1.3642  404.178711  ...          0          0      1
15        0.0998    1.3621  404.178711  ...          0          0      1
...          ...       ...         ...  ...        ...        ...    ...
181782    0.0305    1.3042  392.330292  ...          0          0      1
181787    0.0305    1.3270  392.330292  ...          0          0      1
181791    0.0305    1.3232  392.330292  ...          0          0      1
181792    0.0305    1.3269  392.330292  ...          0          0      1
181797    0.0305    1.3254  392.330292  ...          0          0      1

[45866 rows x 64 columns]

time: 2min 11s


# **A1_A2**

5

In [19]:
picklename = '/drive/My Drive/data/AAN_new/A1_A2/reset_Error_whole_x_test_5percent.pickle'
percent = 0.05
target = 2
max = 500 
target2 = 3
max2 = 1100

# A1: 2 / 500
# A2: 3 / 1100
# N:  4 / 32

avg_ratio = 0.1     # 평균의 반영비율
baesu = 1           # 평균*배수. 배수 설정.
sector_size = 200
sector_num = 909     

x_test = pd.read_pickle(os.getcwd() + '/drive/My Drive/data/AAN_new/reset_NoError_whole_x_test.pickle')
x_test.drop('time', axis=1, inplace=True)
 
for i in range(0, sector_num):
  sector_start = i * sector_size
  sector_end = sector_start + sector_size

  temp = x_test.iloc[sector_start:sector_end]   # 섹터추출
  normal_index = temp.loc[(temp.attack == 0)].index # 섹터에서 normal인 index 추출
  rand_index = random.sample(list(normal_index), int(len(normal_index)*percent))  # normal갯수*percent의 올림갯수만큼 random선택.   

  # 타겟에 더할 값 계산. avg = 평균*비율. avg는 평균이 아님.
  avg = temp.iloc[:, target].mean(axis=0) * avg_ratio
  avg2 = temp.iloc[:, target2].mean(axis=0) * avg_ratio
  

  # random index를 가지고 x_test에 에러 넣기
  for j in (rand_index):
    new_value = 0
    times = baesu
    while(new_value < max + avg):
      new_value = x_test.iloc[j, target] + (avg*times)    # 에러의 세기 충분하도록 함
      times += 0.3

    new_value2 = 0
    times = baesu
    while(new_value2 < max2 + avg2):
      new_value2 = x_test.iloc[j, target2] + (avg2*times)    # 에러의 세기 충분하도록 함
      times += 0.3
    
    x_test.iloc[j, target] = new_value  # 에러가 들어간 값으로 업데이트
    x_test.iloc[j, target2] = new_value2  # 에러가 들어간 값으로 업데이트
    x_test.iloc[j, 63] = 1  # 에러플래그 1로

############################################### 저장 구간 ##################################################
# save as pickle here.
x_test.to_pickle(os.getcwd() + picklename)
print("pickle saved.")

# pickle test
a = pd.read_pickle(os.getcwd() + picklename)
print("")
print("로드된 데이터")
num0=0
num1=0
for i in a["error"]:
  if(i==0):
    num0 = num0+1
  elif(i==1):
    num1 = num1+1

print('에러 X')
print(num0)
print('에러 O')
print(num1)


# Get names of indexes for which column Age has value 30
indexNames = a[a['error'] == 0 ].index
# Delete these row indexes from dataFrame
a.drop(indexNames , inplace=True)
a


pickle saved.

로드된 데이터
에러 X
174156
에러 O
7644


P1.B2004  P1.B2016    P1.B3004  ...  attack.P2  attack.P3  error
15        0.0998    1.3621  541.599731  ...          0          0      1
37        0.0998    1.3247  541.599731  ...          0          0      1
53        0.0998    1.3393  541.599731  ...          0          0      1
65        0.0998    1.4480  541.599731  ...          0          0      1
66        0.0998    1.4510  541.599731  ...          0          0      1
...          ...       ...         ...  ...        ...        ...    ...
181703    0.0305    1.1629  549.262512  ...          0          0      1
181715    0.0305    1.0988  549.262512  ...          0          0      1
181725    0.0305    1.0795  549.262512  ...          0          0      1
181748    0.0305    1.3084  549.262512  ...          0          0      1
181794    0.0305    1.3209  549.262512  ...          0          0      1

[7644 rows x 64 columns]

time: 25.1 s


10

In [20]:
picklename = '/drive/My Drive/data/AAN_new/A1_A2/reset_Error_whole_x_test_10percent.pickle'
percent = 0.1
target = 2
max = 500 
target2 = 3
max2 = 1100

# A1: 2 / 500
# A2: 3 / 1100
# N:  4 / 32

avg_ratio = 0.1     # 평균의 반영비율
baesu = 1           # 평균*배수. 배수 설정.
sector_size = 200
sector_num = 909     

x_test = pd.read_pickle(os.getcwd() + '/drive/My Drive/data/AAN_new/reset_NoError_whole_x_test.pickle')
x_test.drop('time', axis=1, inplace=True)
 
for i in range(0, sector_num):
  sector_start = i * sector_size
  sector_end = sector_start + sector_size

  temp = x_test.iloc[sector_start:sector_end]   # 섹터추출
  normal_index = temp.loc[(temp.attack == 0)].index # 섹터에서 normal인 index 추출
  rand_index = random.sample(list(normal_index), int(len(normal_index)*percent))  # normal갯수*percent의 올림갯수만큼 random선택.   

  # 타겟에 더할 값 계산. avg = 평균*비율. avg는 평균이 아님.
  avg = temp.iloc[:, target].mean(axis=0) * avg_ratio
  avg2 = temp.iloc[:, target2].mean(axis=0) * avg_ratio
  

  # random index를 가지고 x_test에 에러 넣기
  for j in (rand_index):
    new_value = 0
    times = baesu
    while(new_value < max + avg):
      new_value = x_test.iloc[j, target] + (avg*times)    # 에러의 세기 충분하도록 함
      times += 0.3

    new_value2 = 0
    times = baesu
    while(new_value2 < max2 + avg2):
      new_value2 = x_test.iloc[j, target2] + (avg2*times)    # 에러의 세기 충분하도록 함
      times += 0.3
    
    x_test.iloc[j, target] = new_value  # 에러가 들어간 값으로 업데이트
    x_test.iloc[j, target2] = new_value2  # 에러가 들어간 값으로 업데이트
    x_test.iloc[j, 63] = 1  # 에러플래그 1로

############################################### 저장 구간 ##################################################
# save as pickle here.
x_test.to_pickle(os.getcwd() + picklename)
print("pickle saved.")

# pickle test
a = pd.read_pickle(os.getcwd() + picklename)
print("")
print("로드된 데이터")
num0=0
num1=0
for i in a["error"]:
  if(i==0):
    num0 = num0+1
  elif(i==1):
    num1 = num1+1

print('에러 X')
print(num0)
print('에러 O')
print(num1)


# Get names of indexes for which column Age has value 30
indexNames = a[a['error'] == 0 ].index
# Delete these row indexes from dataFrame
a.drop(indexNames , inplace=True)
a


pickle saved.

로드된 데이터
에러 X
166512
에러 O
15288


P1.B2004  P1.B2016    P1.B3004  ...  attack.P2  attack.P3  error
2         0.0998    1.3837  541.599731  ...          0          0      1
8         0.0998    1.3723  541.599731  ...          0          0      1
11        0.0998    1.3647  541.599731  ...          0          0      1
18        0.0998    1.3528  541.599731  ...          0          0      1
23        0.0998    1.3456  541.599731  ...          0          0      1
...          ...       ...         ...  ...        ...        ...    ...
181735    0.0305    1.1796  549.262512  ...          0          0      1
181744    0.0305    1.2688  549.262512  ...          0          0      1
181751    0.0305    1.3339  549.262512  ...          0          0      1
181775    0.0305    1.3383  549.262512  ...          0          0      1
181786    0.0305    1.3254  549.262512  ...          0          0      1

[15288 rows x 64 columns]

time: 47.5 s


15

In [21]:
picklename = '/drive/My Drive/data/AAN_new/A1_A2/reset_Error_whole_x_test_15percent.pickle'
percent = 0.15
target = 2
max = 500 
target2 = 3
max2 = 1100

# A1: 2 / 500
# A2: 3 / 1100
# N:  4 / 32

avg_ratio = 0.1     # 평균의 반영비율
baesu = 1           # 평균*배수. 배수 설정.
sector_size = 200
sector_num = 909     

x_test = pd.read_pickle(os.getcwd() + '/drive/My Drive/data/AAN_new/reset_NoError_whole_x_test.pickle')
x_test.drop('time', axis=1, inplace=True)
 
for i in range(0, sector_num):
  sector_start = i * sector_size
  sector_end = sector_start + sector_size

  temp = x_test.iloc[sector_start:sector_end]   # 섹터추출
  normal_index = temp.loc[(temp.attack == 0)].index # 섹터에서 normal인 index 추출
  rand_index = random.sample(list(normal_index), int(len(normal_index)*percent))  # normal갯수*percent의 올림갯수만큼 random선택.   

  # 타겟에 더할 값 계산. avg = 평균*비율. avg는 평균이 아님.
  avg = temp.iloc[:, target].mean(axis=0) * avg_ratio
  avg2 = temp.iloc[:, target2].mean(axis=0) * avg_ratio
  

  # random index를 가지고 x_test에 에러 넣기
  for j in (rand_index):
    new_value = 0
    times = baesu
    while(new_value < max + avg):
      new_value = x_test.iloc[j, target] + (avg*times)    # 에러의 세기 충분하도록 함
      times += 0.3

    new_value2 = 0
    times = baesu
    while(new_value2 < max2 + avg2):
      new_value2 = x_test.iloc[j, target2] + (avg2*times)    # 에러의 세기 충분하도록 함
      times += 0.3
    
    x_test.iloc[j, target] = new_value  # 에러가 들어간 값으로 업데이트
    x_test.iloc[j, target2] = new_value2  # 에러가 들어간 값으로 업데이트
    x_test.iloc[j, 63] = 1  # 에러플래그 1로

############################################### 저장 구간 ##################################################
# save as pickle here.
x_test.to_pickle(os.getcwd() + picklename)
print("pickle saved.")

# pickle test
a = pd.read_pickle(os.getcwd() + picklename)
print("")
print("로드된 데이터")
num0=0
num1=0
for i in a["error"]:
  if(i==0):
    num0 = num0+1
  elif(i==1):
    num1 = num1+1

print('에러 X')
print(num0)
print('에러 O')
print(num1)


# Get names of indexes for which column Age has value 30
indexNames = a[a['error'] == 0 ].index
# Delete these row indexes from dataFrame
a.drop(indexNames , inplace=True)
a


pickle saved.

로드된 데이터
에러 X
158867
에러 O
22933


P1.B2004  P1.B2016    P1.B3004  ...  attack.P2  attack.P3  error
7         0.0998    1.3746  541.599731  ...          0          0      1
9         0.0998    1.3711  541.599731  ...          0          0      1
10        0.0998    1.3688  541.599731  ...          0          0      1
15        0.0998    1.3621  541.599731  ...          0          0      1
17        0.0998    1.3564  541.599731  ...          0          0      1
...          ...       ...         ...  ...        ...        ...    ...
181758    0.0305    1.3574  549.262512  ...          0          0      1
181763    0.0305    1.3859  549.262512  ...          0          0      1
181772    0.0305    1.3560  549.262512  ...          0          0      1
181785    0.0305    1.3176  549.262512  ...          0          0      1
181799    0.0305    1.3251  549.262512  ...          0          0      1

[22933 rows x 64 columns]

time: 1min 7s


30

In [22]:
picklename = '/drive/My Drive/data/AAN_new/A1_A2/reset_Error_whole_x_test_30percent.pickle'
percent = 0.3
target = 2
max = 500 
target2 = 3
max2 = 1100

# A1: 2 / 500
# A2: 3 / 1100
# N:  4 / 32

avg_ratio = 0.1     # 평균의 반영비율
baesu = 1           # 평균*배수. 배수 설정.
sector_size = 200
sector_num = 909     

x_test = pd.read_pickle(os.getcwd() + '/drive/My Drive/data/AAN_new/reset_NoError_whole_x_test.pickle')
x_test.drop('time', axis=1, inplace=True)
 
for i in range(0, sector_num):
  sector_start = i * sector_size
  sector_end = sector_start + sector_size

  temp = x_test.iloc[sector_start:sector_end]   # 섹터추출
  normal_index = temp.loc[(temp.attack == 0)].index # 섹터에서 normal인 index 추출
  rand_index = random.sample(list(normal_index), int(len(normal_index)*percent))  # normal갯수*percent의 올림갯수만큼 random선택.   

  # 타겟에 더할 값 계산. avg = 평균*비율. avg는 평균이 아님.
  avg = temp.iloc[:, target].mean(axis=0) * avg_ratio
  avg2 = temp.iloc[:, target2].mean(axis=0) * avg_ratio
  

  # random index를 가지고 x_test에 에러 넣기
  for j in (rand_index):
    new_value = 0
    times = baesu
    while(new_value < max + avg):
      new_value = x_test.iloc[j, target] + (avg*times)    # 에러의 세기 충분하도록 함
      times += 0.3

    new_value2 = 0
    times = baesu
    while(new_value2 < max2 + avg2):
      new_value2 = x_test.iloc[j, target2] + (avg2*times)    # 에러의 세기 충분하도록 함
      times += 0.3
    
    x_test.iloc[j, target] = new_value  # 에러가 들어간 값으로 업데이트
    x_test.iloc[j, target2] = new_value2  # 에러가 들어간 값으로 업데이트
    x_test.iloc[j, 63] = 1  # 에러플래그 1로

############################################### 저장 구간 ##################################################
# save as pickle here.
x_test.to_pickle(os.getcwd() + picklename)
print("pickle saved.")

# pickle test
a = pd.read_pickle(os.getcwd() + picklename)
print("")
print("로드된 데이터")
num0=0
num1=0
for i in a["error"]:
  if(i==0):
    num0 = num0+1
  elif(i==1):
    num1 = num1+1

print('에러 X')
print(num0)
print('에러 O')
print(num1)


# Get names of indexes for which column Age has value 30
indexNames = a[a['error'] == 0 ].index
# Delete these row indexes from dataFrame
a.drop(indexNames , inplace=True)
a


pickle saved.

로드된 데이터
에러 X
135934
에러 O
45866


P1.B2004  P1.B2016    P1.B3004  ...  attack.P2  attack.P3  error
2         0.0998    1.3837  541.599731  ...          0          0      1
3         0.0998    1.3854  541.599731  ...          0          0      1
6         0.0998    1.3785  541.599731  ...          0          0      1
7         0.0998    1.3746  541.599731  ...          0          0      1
8         0.0998    1.3723  541.599731  ...          0          0      1
...          ...       ...         ...  ...        ...        ...    ...
181783    0.0305    1.3083  549.262512  ...          0          0      1
181786    0.0305    1.3254  549.262512  ...          0          0      1
181787    0.0305    1.3270  549.262512  ...          0          0      1
181792    0.0305    1.3269  549.262512  ...          0          0      1
181796    0.0305    1.3258  549.262512  ...          0          0      1

[45866 rows x 64 columns]

time: 2min 12s


# **A1 A2 N**

5

In [23]:
picklename = '/drive/My Drive/data/AAN_new/A1_A2_N/reset_Error_whole_x_test_5percent.pickle'
percent = 0.05
target = 2
max = 500 
target2 = 3
max2 = 1100
target3 = 4
max3 = 32

# A1: 2 / 500
# A2: 3 / 1100
# N:  4 / 32

avg_ratio = 0.1     # 평균의 반영비율
baesu = 1           # 평균*배수. 배수 설정.
sector_size = 200
sector_num = 909     

x_test = pd.read_pickle(os.getcwd() + '/drive/My Drive/data/AAN_new/reset_NoError_whole_x_test.pickle')
x_test.drop('time', axis=1, inplace=True)
 
for i in range(0, sector_num):
  sector_start = i * sector_size
  sector_end = sector_start + sector_size

  temp = x_test.iloc[sector_start:sector_end]   # 섹터추출
  normal_index = temp.loc[(temp.attack == 0)].index # 섹터에서 normal인 index 추출
  rand_index = random.sample(list(normal_index), int(len(normal_index)*percent))  # normal갯수*percent의 올림갯수만큼 random선택.   

  # 타겟에 더할 값 계산. avg = 평균*비율. avg는 평균이 아님.
  avg = temp.iloc[:, target].mean(axis=0) * avg_ratio
  avg2 = temp.iloc[:, target2].mean(axis=0) * avg_ratio
  avg3 = temp.iloc[:, target3].mean(axis=0) * avg_ratio
  
  # random index를 가지고 x_test에 에러 넣기
  for j in (rand_index):
    new_value = 0
    times = baesu
    while(new_value < max + avg):
      new_value = x_test.iloc[j, target] + (avg*times)    # 에러의 세기 충분하도록 함
      times += 0.3

    new_value2 = 0
    times = baesu
    while(new_value2 < max2 + avg2):
      new_value2 = x_test.iloc[j, target2] + (avg2*times)    # 에러의 세기 충분하도록 함
      times += 0.3

    new_value3 = 0
    times = baesu
    while(new_value3 < max3 + avg3):
      new_value3 = x_test.iloc[j, target3] + (avg3*times)    # 에러의 세기 충분하도록 함
      times += 0.3
    
    x_test.iloc[j, target] = new_value  # 에러가 들어간 값으로 업데이트
    x_test.iloc[j, target2] = new_value2  # 에러가 들어간 값으로 업데이트
    x_test.iloc[j, target3] = new_value3  # 에러가 들어간 값으로 업데이트
    x_test.iloc[j, 63] = 1  # 에러플래그 1로

############################################### 저장 구간 ##################################################
# save as pickle here.
x_test.to_pickle(os.getcwd() + picklename)
print("pickle saved.")

# pickle test
a = pd.read_pickle(os.getcwd() + picklename)
print("")
print("로드된 데이터")
num0=0
num1=0
for i in a["error"]:
  if(i==0):
    num0 = num0+1
  elif(i==1):
    num1 = num1+1

print('에러 X')
print(num0)
print('에러 O')
print(num1)


# Get names of indexes for which column Age has value 30
indexNames = a[a['error'] == 0 ].index
# Delete these row indexes from dataFrame
a.drop(indexNames , inplace=True)
a


pickle saved.

로드된 데이터
에러 X
174156
에러 O
7644


P1.B2004  P1.B2016    P1.B3004  ...  attack.P2  attack.P3  error
8         0.0998    1.3723  541.599731  ...          0          0      1
15        0.0998    1.3621  541.599731  ...          0          0      1
57        0.0998    1.3764  541.599731  ...          0          0      1
63        0.0998    1.4254  541.599731  ...          0          0      1
99        0.0998    1.6321  541.599731  ...          0          0      1
...          ...       ...         ...  ...        ...        ...    ...
181683    0.0305    1.2619  549.262512  ...          0          0      1
181688    0.0305    1.2482  549.262512  ...          0          0      1
181690    0.0305    1.2445  549.262512  ...          0          0      1
181691    0.0305    1.2331  549.262512  ...          0          0      1
181766    0.0305    1.3992  549.262512  ...          0          0      1

[7644 rows x 64 columns]

time: 31.1 s


10

In [24]:
picklename = '/drive/My Drive/data/AAN_new/A1_A2_N/reset_Error_whole_x_test_10percent.pickle'
percent = 0.1
target = 2
max = 500 
target2 = 3
max2 = 1100
target3 = 4
max3 = 32

# A1: 2 / 500
# A2: 3 / 1100
# N:  4 / 32

avg_ratio = 0.1     # 평균의 반영비율
baesu = 1           # 평균*배수. 배수 설정.
sector_size = 200
sector_num = 909     

x_test = pd.read_pickle(os.getcwd() + '/drive/My Drive/data/AAN_new/reset_NoError_whole_x_test.pickle')
x_test.drop('time', axis=1, inplace=True)
 
for i in range(0, sector_num):
  sector_start = i * sector_size
  sector_end = sector_start + sector_size

  temp = x_test.iloc[sector_start:sector_end]   # 섹터추출
  normal_index = temp.loc[(temp.attack == 0)].index # 섹터에서 normal인 index 추출
  rand_index = random.sample(list(normal_index), int(len(normal_index)*percent))  # normal갯수*percent의 올림갯수만큼 random선택.   

  # 타겟에 더할 값 계산. avg = 평균*비율. avg는 평균이 아님.
  avg = temp.iloc[:, target].mean(axis=0) * avg_ratio
  avg2 = temp.iloc[:, target2].mean(axis=0) * avg_ratio
  avg3 = temp.iloc[:, target3].mean(axis=0) * avg_ratio
  
  # random index를 가지고 x_test에 에러 넣기
  for j in (rand_index):
    new_value = 0
    times = baesu
    while(new_value < max + avg):
      new_value = x_test.iloc[j, target] + (avg*times)    # 에러의 세기 충분하도록 함
      times += 0.3

    new_value2 = 0
    times = baesu
    while(new_value2 < max2 + avg2):
      new_value2 = x_test.iloc[j, target2] + (avg2*times)    # 에러의 세기 충분하도록 함
      times += 0.3

    new_value3 = 0
    times = baesu
    while(new_value3 < max3 + avg3):
      new_value3 = x_test.iloc[j, target3] + (avg3*times)    # 에러의 세기 충분하도록 함
      times += 0.3
    
    x_test.iloc[j, target] = new_value  # 에러가 들어간 값으로 업데이트
    x_test.iloc[j, target2] = new_value2  # 에러가 들어간 값으로 업데이트
    x_test.iloc[j, target3] = new_value3  # 에러가 들어간 값으로 업데이트
    x_test.iloc[j, 63] = 1  # 에러플래그 1로

############################################### 저장 구간 ##################################################
# save as pickle here.
x_test.to_pickle(os.getcwd() + picklename)
print("pickle saved.")

# pickle test
a = pd.read_pickle(os.getcwd() + picklename)
print("")
print("로드된 데이터")
num0=0
num1=0
for i in a["error"]:
  if(i==0):
    num0 = num0+1
  elif(i==1):
    num1 = num1+1

print('에러 X')
print(num0)
print('에러 O')
print(num1)


# Get names of indexes for which column Age has value 30
indexNames = a[a['error'] == 0 ].index
# Delete these row indexes from dataFrame
a.drop(indexNames , inplace=True)
a


pickle saved.

로드된 데이터
에러 X
166512
에러 O
15288


P1.B2004  P1.B2016    P1.B3004  ...  attack.P2  attack.P3  error
5         0.0998    1.3804  541.599731  ...          0          0      1
29        0.0998    1.3393  541.599731  ...          0          0      1
36        0.0998    1.3274  541.599731  ...          0          0      1
43        0.0998    1.3166  541.599731  ...          0          0      1
78        0.0998    1.5459  541.599731  ...          0          0      1
...          ...       ...         ...  ...        ...        ...    ...
181757    0.0305    1.3577  549.262512  ...          0          0      1
181771    0.0305    1.3661  549.262512  ...          0          0      1
181785    0.0305    1.3176  549.262512  ...          0          0      1
181792    0.0305    1.3269  549.262512  ...          0          0      1
181799    0.0305    1.3251  549.262512  ...          0          0      1

[15288 rows x 64 columns]

time: 58.6 s


15

In [25]:
picklename = '/drive/My Drive/data/AAN_new/A1_A2_N/reset_Error_whole_x_test_15percent.pickle'
percent = 0.15
target = 2
max = 500 
target2 = 3
max2 = 1100
target3 = 4
max3 = 32

# A1: 2 / 500
# A2: 3 / 1100
# N:  4 / 32

avg_ratio = 0.1     # 평균의 반영비율
baesu = 1           # 평균*배수. 배수 설정.
sector_size = 200
sector_num = 909     

x_test = pd.read_pickle(os.getcwd() + '/drive/My Drive/data/AAN_new/reset_NoError_whole_x_test.pickle')
x_test.drop('time', axis=1, inplace=True)
 
for i in range(0, sector_num):
  sector_start = i * sector_size
  sector_end = sector_start + sector_size

  temp = x_test.iloc[sector_start:sector_end]   # 섹터추출
  normal_index = temp.loc[(temp.attack == 0)].index # 섹터에서 normal인 index 추출
  rand_index = random.sample(list(normal_index), int(len(normal_index)*percent))  # normal갯수*percent의 올림갯수만큼 random선택.   

  # 타겟에 더할 값 계산. avg = 평균*비율. avg는 평균이 아님.
  avg = temp.iloc[:, target].mean(axis=0) * avg_ratio
  avg2 = temp.iloc[:, target2].mean(axis=0) * avg_ratio
  avg3 = temp.iloc[:, target3].mean(axis=0) * avg_ratio
  
  # random index를 가지고 x_test에 에러 넣기
  for j in (rand_index):
    new_value = 0
    times = baesu
    while(new_value < max + avg):
      new_value = x_test.iloc[j, target] + (avg*times)    # 에러의 세기 충분하도록 함
      times += 0.3

    new_value2 = 0
    times = baesu
    while(new_value2 < max2 + avg2):
      new_value2 = x_test.iloc[j, target2] + (avg2*times)    # 에러의 세기 충분하도록 함
      times += 0.3

    new_value3 = 0
    times = baesu
    while(new_value3 < max3 + avg3):
      new_value3 = x_test.iloc[j, target3] + (avg3*times)    # 에러의 세기 충분하도록 함
      times += 0.3
    
    x_test.iloc[j, target] = new_value  # 에러가 들어간 값으로 업데이트
    x_test.iloc[j, target2] = new_value2  # 에러가 들어간 값으로 업데이트
    x_test.iloc[j, target3] = new_value3  # 에러가 들어간 값으로 업데이트
    x_test.iloc[j, 63] = 1  # 에러플래그 1로

############################################### 저장 구간 ##################################################
# save as pickle here.
x_test.to_pickle(os.getcwd() + picklename)
print("pickle saved.")

# pickle test
a = pd.read_pickle(os.getcwd() + picklename)
print("")
print("로드된 데이터")
num0=0
num1=0
for i in a["error"]:
  if(i==0):
    num0 = num0+1
  elif(i==1):
    num1 = num1+1

print('에러 X')
print(num0)
print('에러 O')
print(num1)


# Get names of indexes for which column Age has value 30
indexNames = a[a['error'] == 0 ].index
# Delete these row indexes from dataFrame
a.drop(indexNames , inplace=True)
a


pickle saved.

로드된 데이터
에러 X
158867
에러 O
22933


P1.B2004  P1.B2016    P1.B3004  ...  attack.P2  attack.P3  error
0         0.0998    1.3901  541.599731  ...          0          0      1
1         0.0998    1.3840  541.599731  ...          0          0      1
16        0.0998    1.3589  541.599731  ...          0          0      1
21        0.0998    1.3495  541.599731  ...          0          0      1
24        0.0998    1.3413  541.599731  ...          0          0      1
...          ...       ...         ...  ...        ...        ...    ...
181735    0.0305    1.1796  549.262512  ...          0          0      1
181759    0.0305    1.3600  549.262512  ...          0          0      1
181764    0.0305    1.3896  549.262512  ...          0          0      1
181768    0.0305    1.3896  549.262512  ...          0          0      1
181791    0.0305    1.3232  549.262512  ...          0          0      1

[22933 rows x 64 columns]

time: 1min 25s


30

In [26]:
picklename = '/drive/My Drive/data/AAN_new/A1_A2_N/reset_Error_whole_x_test_30percent.pickle'
percent = 0.3
target = 2
max = 500 
target2 = 3
max2 = 1100
target3 = 4
max3 = 32

# A1: 2 / 500
# A2: 3 / 1100
# N:  4 / 32

avg_ratio = 0.1     # 평균의 반영비율
baesu = 1           # 평균*배수. 배수 설정.
sector_size = 200
sector_num = 909     

x_test = pd.read_pickle(os.getcwd() + '/drive/My Drive/data/AAN_new/reset_NoError_whole_x_test.pickle')
x_test.drop('time', axis=1, inplace=True)
 
for i in range(0, sector_num):
  sector_start = i * sector_size
  sector_end = sector_start + sector_size

  temp = x_test.iloc[sector_start:sector_end]   # 섹터추출
  normal_index = temp.loc[(temp.attack == 0)].index # 섹터에서 normal인 index 추출
  rand_index = random.sample(list(normal_index), int(len(normal_index)*percent))  # normal갯수*percent의 올림갯수만큼 random선택.   

  # 타겟에 더할 값 계산. avg = 평균*비율. avg는 평균이 아님.
  avg = temp.iloc[:, target].mean(axis=0) * avg_ratio
  avg2 = temp.iloc[:, target2].mean(axis=0) * avg_ratio
  avg3 = temp.iloc[:, target3].mean(axis=0) * avg_ratio
  
  # random index를 가지고 x_test에 에러 넣기
  for j in (rand_index):
    new_value = 0
    times = baesu
    while(new_value < max + avg):
      new_value = x_test.iloc[j, target] + (avg*times)    # 에러의 세기 충분하도록 함
      times += 0.3

    new_value2 = 0
    times = baesu
    while(new_value2 < max2 + avg2):
      new_value2 = x_test.iloc[j, target2] + (avg2*times)    # 에러의 세기 충분하도록 함
      times += 0.3

    new_value3 = 0
    times = baesu
    while(new_value3 < max3 + avg3):
      new_value3 = x_test.iloc[j, target3] + (avg3*times)    # 에러의 세기 충분하도록 함
      times += 0.3
    
    x_test.iloc[j, target] = new_value  # 에러가 들어간 값으로 업데이트
    x_test.iloc[j, target2] = new_value2  # 에러가 들어간 값으로 업데이트
    x_test.iloc[j, target3] = new_value3  # 에러가 들어간 값으로 업데이트
    x_test.iloc[j, 63] = 1  # 에러플래그 1로

############################################### 저장 구간 ##################################################
# save as pickle here.
x_test.to_pickle(os.getcwd() + picklename)
print("pickle saved.")

# pickle test
a = pd.read_pickle(os.getcwd() + picklename)
print("")
print("로드된 데이터")
num0=0
num1=0
for i in a["error"]:
  if(i==0):
    num0 = num0+1
  elif(i==1):
    num1 = num1+1

print('에러 X')
print(num0)
print('에러 O')
print(num1)


# Get names of indexes for which column Age has value 30
indexNames = a[a['error'] == 0 ].index
# Delete these row indexes from dataFrame
a.drop(indexNames , inplace=True)
a


pickle saved.

로드된 데이터
에러 X
135934
에러 O
45866


P1.B2004  P1.B2016    P1.B3004  ...  attack.P2  attack.P3  error
0         0.0998    1.3901  541.599731  ...          0          0      1
2         0.0998    1.3837  541.599731  ...          0          0      1
3         0.0998    1.3854  541.599731  ...          0          0      1
16        0.0998    1.3589  541.599731  ...          0          0      1
17        0.0998    1.3564  541.599731  ...          0          0      1
...          ...       ...         ...  ...        ...        ...    ...
181786    0.0305    1.3254  549.262512  ...          0          0      1
181788    0.0305    1.3252  549.262512  ...          0          0      1
181790    0.0305    1.3263  549.262512  ...          0          0      1
181793    0.0305    1.3283  549.262512  ...          0          0      1
181797    0.0305    1.3254  549.262512  ...          0          0      1

[45866 rows x 64 columns]

time: 2min 52s
